---
title: 'PLAE enables quick and easy searching across one million ocular transcriptomes'
author:
  - Vinay Swamy:
      institute: bg
  - Zachary Batz:
      institute: nnrl
  - Robert Hufnagel:
      institute: mgog
  - David McGaughey:
      institute:
        - bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, National Institutes of Health
  - nnrl: Neurobiology, Neurodegeneration & Repair Laboratory, National Eye Institute, National Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health
```{r}

```
    
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  word_document:
    reference_docx: word-styles-reference-01.docx
    fig_caption: yes
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
bibliography: zotero_plaeApp.bib
csl: investigative-ophthalmology-and-visual-science.csl
abstract: "PURPOSE: To create a high performance reactive web application to query gene expression across cell type, species, study, and other factors. METHODS: We wrote the Shiny web application PLAE (https://plae.nei.nih.gov) and tailored the structure of the underlying data (single cell Eye in a Disk, scEiaD) to asdfasdf. RESULTS: The new portal provides quick visualization of over a million individual cells from vertebrate eye-and body transcriptomes., XX cell types, XX ocular tissues, XX body tissues. As a test of the value of this unified pan-eye dataset, we show XX. CONCLUSION: The PLAE v0.80 web app serves the pan-ocular and body dataset, scEiaD. This offers the eye community a powerful and quick means to test hypotheses on gene and transcript expression across 54 body and 19 eye tissues. "
keywords: "RNA-seq, lens, cornea, retina, RPE, Snakemake, web, app"
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)

# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

library(pool)
library(RSQLite)
meta_filter_v0 <- fst::read_fst("~/data/scEiaD/2021_03_17_meta_filter.fst") %>% 
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType))
# remove eric bharti 2D/3D RPE
meta_filter <- fst::read_fst("/Volumes/McGaughey_S/scEiaD_v3/2021_11_09_meta_filter.fst") %>% filter(study_accession != 'Bharti_Nguyen_iRPE_2D_3D') %>% 
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType))

```

# (sup?) Table 1
Studies in scEiaD
```{r}
meta_filter %>% group_by(Citation, PMID, study_accession) %>% summarise(Count = n()) %>% flextable()
```

# Figure 1 Distribution of Tissues and Cell Types in scEiaD
```{r, fig.width=8, fig.height=5}
# organ
meta_filter <- meta_filter %>% mutate(organism = factor(organism, levels = meta_filter$organism %>% unique()))
a <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Body') %>% 
  filter(Tissue != 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  # scale_y_continuous(breaks = c(2500, 5000, 7500, 10000, 12500, 84000)) +
  # ggbreak::scale_y_break(c(13000, 82000)) +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

aa <- meta_filter %>% 
  filter(Tissue == 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

# eye studies
b <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = case_when(Tissue == "Choroid" ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  group_by(Organ, Tissue,organism) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) +
  xlab('')

# eye studies samples
c <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  select(Organ, Tissue,organism, study_accession) %>%
  unique() %>%
  group_by(Organ, Tissue, organism) %>% 
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0))  + xlab('')

# retina cell types
d <- meta_filter %>% filter(Organ == 'Eye') %>% 
  group_by(CellType_predict, study_accession, organism) %>%
  summarise(Count = n()) %>% filter(Count > 10) %>% ungroup() %>% 
  ggplot(aes(x=CellType_predict,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')


a_aa_b <- plot_grid(a,aa,b, ncol = 1, align = 'hv', rel_heights = c(10, 1.5,3.5), labels = 'a')
c_d <- plot_grid(c,d, rel_heights = c(2,10), ncol = 1, align= 'hv', axis = 'l',labels = c('b','c'))
plot_grid(a_aa_b, c_d, ncol = 2, rel_widths = c(1,1.2))
```

# Sup fig 1 
DecontX *in silico* ambient RNA contamination tool substantially removes Rhodopsin expression in non-rod cells while retaining high expression rods. First shown (a) is Rhodopsin expression across the scEiaD v0 dataset, without DecontX optimization. Note Rhodopsin expression is noticeable in many cell types beyond the rods. Next (b) is the scEiaD v1 dataset with DecontX optimization. Rhodopsin is nearly exclusively expressed in labelled rod cells. Despite the computational ambient RNA removal, Rhodopsin expression remains high in the Rods. 

```{r, fig.width=10, fig.height=12}
source('src/make_exp_plot.R')
scEiaD_2020_v00 <- pool::dbPool(RSQLite::SQLite(), dbname =  '~/data/scEiaD/MOARTABLES__anthology_limmaFALSE___5000-transform-counts-universe-batch-scVIprojectionSO-8-0.1-50-5.sqlite')
scEiaD_2020_v01 <- pool::dbPool(RSQLite::SQLite(), dbname =  '/Volumes/McGaughey_S/scEiaD_v3/MOARTABLES__anthology_limmaFALSE___6000-counts-universe-batch-scVIprojection-10-15-0.1-50-20.sqlite')

input <- list()
input$exp_plot_groups <- c('organism')
input$exp_plot_facet <- c('CellType_predict')
input$exp_plot_genes <- c('RHO (ENSG00000163914)')
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 50

v0_rho <- make_exp_plot(input, scEiaD_2020_v00, meta_filter_v0)
v1_rho <-  make_exp_plot(input, scEiaD_2020_v01, meta_filter)
input$exp_filter_min_cell_number <- 1000
v1.1_rho <- make_exp_plot(input, scEiaD_2020_v01, meta_filter %>% filter(CellType_predict %in% unique(meta_filter_v0$CellType_predict)))
plot_grid(v0_rho, v1_rho, ncol = 1, labels = c('a','b'), rel_heights = c(1,1.5))
```
<!-- ```{r, fig.height=5, fig.width=3} -->
<!-- # Distribution of ML Retina Cell Types Between Mouse and Human in Pan-Retina Studies -->
<!-- Drop this figure? -->
<!-- Yes -->

<!-- pan_retina_studies <- c('SRP255195','E-MTAB-7316','SRP151023','SRP222001','SRP222958','SRP194595','SRP223254','SRP050054','SRP269634','SRP269635','SRP158528', 'SRP257758','SRP212788') -->


<!-- meta_filter %>%  -->
<!--   filter(organism %in% c('Mus musculus','Homo sapiens'),  -->
<!--          Stage == 'Mature',  -->
<!--          study_accession %in% pan_retina_studies,  -->
<!--          grepl('Amacr|Bipol|Cone|Epithel|Endoth|Horizon|Muller|Pericy|Ganglion|Rod', CellType_predict)) %>%  -->
<!--   group_by(study_accession, CellType_predict) %>%  -->
<!--   summarise(Count = n()) %>% mutate(Percentage = 100 * (Count / sum(Count)))  %>%  -->
<!--   left_join(meta_filter %>% select(study_accession, Platform, organism) %>% unique()) %>%  -->
<!--   ggplot(aes(x=organism, y=Percentage)) +  -->
<!--   geom_boxplot() +  -->
<!--   coord_flip() +  -->
<!--   ggbeeswarm::geom_quasirandom(aes(color=study_accession)) +  -->
<!--   facet_wrap(~CellType_predict, ncol = 1) +  -->
<!--   cowplot::theme_cowplot() -->
<!-- ``` -->

# Introduction

## Many published ocular atlases

## Several web atlases with ocular cells exist but have significant limitations

There are several web app resources for querying gene expression across ocular cell types (table 1). 

## Data difficult to access and cross comparisons impossible

But important as seeing trends trends across different groups strongly demonstrates that 
the signal is biological, not technical. 

## What we did

Collated XX studies and processed 

# Methods

## pan mouse and pan human as non-ocular reference
Tabula Muris and 

# Results

## Updates to the scEiaD backend
Version zero of the single cell Eye is a Disk (scEiaD) database contained `r meta_filter_v0$study_accession %>% unique() %>% length()` studies, three species, `r nrow(meta_filter_v0)` cells, and `r meta_filter_v0$CellType_predict %>% unique() %>% length()` curated cell types. We updated the scEiaD resource for the PLAE v0.80 web app in five important ways. First, we added two chicken retina (*Gallus gallus*) studies. Second we added human ocular outflow tract and cornea datasets. Third, we added a human pan-body reference scRNA dataset. Fourth, we used an *in silico* method to remove background gene contamination as we noticed persistent Rhodopsin expression in many non-photoreceptors cells (sup fig XX). Fifth, we increased the minimum number of detected unique transcripts per cell required for a cell to be retained from 300 to 450. Sixth, we carried and harmonized several common cell type labels from the Tabula Muris dataset to more easily provide non eye cell type comparisons. 

After the dataset and quality control updates the scEiaD v02 dataset now contains `r meta_filter$study_accession %>% unique() %>% length()` studies, four species, `r nrow(meta_filter)` cells, and `r meta_filter$CellType_predict %>% unique() %>% length()` curated cell types (Fig 1).

# Comparison of PLAE to "pan" scRNA web apps containing ocular data
```{r}
feature_table <- rbind(
  c('PLAE','Yes','Yes', 'Yes', '887', 'https://plae.nei.nih.gov'),
  c('Spectacle', 'No' ,'No', 'No', '87', 'https://singlecell-eye.org'),
  c('UCSC Cell Browser', 'Yes', 'No', 'Yes', '62', 'https://cells.ucsc.edu' ),
  c('Tabula Sapiens', 'No', 'N/A', 'Yes', '10', 'https://tabula-sapiens-portal.ds.czbiohub.org'),
  c('PanglaoDB', 'Yes', 'No', 'Yes', '47', 'https://panglaodb.se'),
  c('Protein Atlas', 'Yes*','No', 'Yes', '5',  'https://www.proteinatlas.org/humanproteome/celltype')) %>% 
  as_tibble()
colnames(feature_table) <- c('Resource Name', 'Independent Datasets', 'Harmonization', 'Ocular and non-Ocular Data', 'Ocular Cell Count\n(thousands)', 'URL')
feature_table %>% flextable()
# as of 2021 10 25
```

## PLAE contains over 10 fold more cells and studies than the closest comparable web app

The Spectacle web app also provides a web app to access 87,305 single cell transcriptomes across four publications.

## Extensive optimization makes the PLAE web app highly responsive despite containing huge amounts of data

## Rich web functionality 

## Differential testing identifies thousands of genes enriched in various ocular cell types

## 

# Conclusion