---
title: 'PLAE enables quick and easy searching across one million ocular transcriptomes'
author:
  - Vinay Swamy:
      institute: bg
  - Zachary Batz:
      institute: nnrl
  - Temesgen Fufa:
      institute: mgog
  - Robert Hufnagel:
      institute: mgog
  - David McGaughey:
      institute:
        - bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, National Institutes of Health
  - nnrl: Neurobiology, Neurodegeneration & Repair Laboratory, National Eye Institute, National Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health
    
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  word_document:
    reference_docx: word-styles-reference-01.docx
    fig_caption: yes
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
bibliography: massive_integrated_eye_scRNA.bib
csl: investigative-ophthalmology-and-visual-science.csl
abstract: "PURPOSE: To create a high performance reactive web application to query gene expression across cell type, species, study, and other factors. METHODS: We wrote the Shiny web application PLAE (https://plae.nei.nih.gov) and tailored the structure of the underlying data (sc Eye in a Disk, scEiaD) to display scEiaD. RESULTS: The new portal provides quick visualization of vertebrate eye-related transcriptomes published to date by gene/transcript, XX cell types, XX ocular tissues, XX body tissues. As a test of the value of this unified pan-eye dataset, we show XX. CONCLUSION: The PLAE v1.0 web app serves the pan-ocular and body dataset, scEiaD. This offers the eye community a powerful and quick means to test hypotheses on human gene and transcript expression across 54 body and 19 eye tissues. "
keywords: "RNA-seq, lens, cornea, retina, RPE, Snakemake, web, app"
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)

# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

```


# Introduction
## Many published ocular atlases

## Data not easily available for all

## Some have accompanying apps

## But apps are generally slow and have minimal exploration tooling

## Cross comparisons difficult
But important as seeing trends trends across different groups strongly demonstrates that 
the signal is biological, not technical. 
  
## What we did

Built an awesome app.

## Will do???

Trajectory

Velocity


# Methods

## Use scEiaD resource
We built, published [].

## pan mouse as non-ocular reference
Tabula Muris

# Results

## Rich functionality

## Web accessible

## 
  
# Conclusion