---
title: 'PLAE enables quick and easy searching across one million ocular transcriptomes'
author:
  - Vinay Swamy:
      institute: bg
  - Zachary Batz:
      institute: nnrl
  - Robert Hufnagel:
      institute: mgog
  - David McGaughey:
      institute:
        - bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, National Institutes of Health
  - nnrl: Neurobiology, Neurodegeneration & Repair Laboratory, National Eye Institute, National Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health
    
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  word_document:
    reference_docx: word-styles-reference-01.docx
    fig_caption: yes
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
bibliography: massive_integrated_eye_scRNA.bib
csl: investigative-ophthalmology-and-visual-science.csl
abstract: "PURPOSE: To create a high performance reactive web application to query gene expression across cell type, species, study, and other factors. METHODS: We wrote the Shiny web application PLAE (https://plae.nei.nih.gov) and tailored the structure of the underlying data (sc Eye in a Disk, scEiaD) to display scEiaD. RESULTS: The new portal provides quick visualization of over a million individual cells from vertebrate eye-and body transcriptomes., XX cell types, XX ocular tissues, XX body tissues. As a test of the value of this unified pan-eye dataset, we show XX. CONCLUSION: The PLAE v0.80 web app serves the pan-ocular and body dataset, scEiaD. This offers the eye community a powerful and quick means to test hypotheses on gene and transcript expression across 54 body and 19 eye tissues. "
keywords: "RNA-seq, lens, cornea, retina, RPE, Snakemake, web, app"
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)

# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

library(pool)
library(RSQLite)
meta_filter <- fst::read_fst("~/data/scEiaD_v2/2021_10_15_meta_filter.fst")
```

# Distribution of Tissues and Cell Types in scEiaD
```{r, fig.width=8, fig.height=5}
# organ
a <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  facet_wrap(~Organ, scales = 'free') +
  coord_flip()

# retina cell types
b <- meta_filter %>% filter(Organ == 'Eye') %>% 
  group_by(CellType_predict, study_accession, organism) %>%
  summarise(Count = n()) %>% filter(Count > 10) %>% ungroup() %>% 
  ggplot(aes(x=CellType_predict,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip()

plot_grid(a,b)
```
# Distribution of ML Retina Cell Types Between Mouse and Human in Pan-Retina Studies
Drop this figure?
```{r, fig.height=5, fig.width=3}
library(tidyverse)
meta_filter <- fst::read_fst('~/data/scEiaD/2021_08_22_meta_filter.fst') %>%
  as_tibble() %>%
  mutate(CellType_predict = case_when(!is.na(TabulaMurisCellType_predict) ~ 'Tabula Muris',
                                      is.na(CellType_predict) ~ 'Unlabelled',
                                      TRUE ~ CellType_predict)) 

pan_retina_studies <- c('SRP255195','E-MTAB-7316','SRP151023','SRP222001','SRP222958','SRP194595','SRP223254','SRP050054','SRP269634','SRP269635','SRP158528', 'SRP257758','SRP212788')


meta_filter %>% 
  filter(organism %in% c('Mus musculus','Homo sapiens'), 
         Stage == 'Mature', 
         study_accession %in% pan_retina_studies, 
         grepl('Amacr|Bipol|Cone|Epithel|Endoth|Horizon|Muller|Pericy|Ganglion|Rod', CellType_predict)) %>% 
  group_by(study_accession, CellType_predict) %>% 
  summarise(Count = n()) %>% mutate(Percentage = 100 * (Count / sum(Count)))  %>% 
  left_join(meta_filter %>% select(study_accession, Platform, organism) %>% unique()) %>% 
  ggplot(aes(x=organism, y=Percentage)) + 
  geom_boxplot() + 
  coord_flip() + 
  ggbeeswarm::geom_quasirandom(aes(color=study_accession)) + 
  facet_wrap(~CellType_predict, ncol = 1) + 
  cowplot::theme_cowplot()
```
# Introduction
## Many published ocular atlases

## Data not easily available for all

## Some have accompanying apps

## But apps are generally slow and have minimal exploration tooling

## Cross comparisons difficult
But important as seeing trends trends across different groups strongly demonstrates that 
the signal is biological, not technical. 

## What we did

Built an awesome app.


# Methods

## Use scEiaD resource
We built, published [].

## pan mouse and pan human as non-ocular reference
Tabula Muris

# Results

## Rich functionality

## Web accessible

## 

# Conclusion