---
title: 'PLAE enables quick and easy searching across one million ocular transcriptomes'
author:
  - Vinay Swamy:
      institute: bg
  - Zachary Batz:
      institute: nnrl
  - Robert Hufnagel:
      institute: mgog
  - David McGaughey:
      institute:
        - bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, National Institutes of Health
  - nnrl: Neurobiology, Neurodegeneration & Repair Laboratory, National Eye Institute, National Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health
    
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  word_document:
    reference_docx: word-styles-reference-01.docx
    fig_caption: yes
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
bibliography: zotero_plaeApp.bib
csl: investigative-ophthalmology-and-visual-science.csl
abstract: "PURPOSE: To create a high performance reactive web application to query gene expression across cell type, species, study, and other factors. METHODS: We wrote the Shiny web application PLAE (https://plae.nei.nih.gov) and tailored the structure of the underlying data (sc Eye in a Disk, scEiaD) to display scEiaD. RESULTS: The new portal provides quick visualization of over a million individual cells from vertebrate eye-and body transcriptomes., XX cell types, XX ocular tissues, XX body tissues. As a test of the value of this unified pan-eye dataset, we show XX. CONCLUSION: The PLAE v0.80 web app serves the pan-ocular and body dataset, scEiaD. This offers the eye community a powerful and quick means to test hypotheses on gene and transcript expression across 54 body and 19 eye tissues. "
keywords: "RNA-seq, lens, cornea, retina, RPE, Snakemake, web, app"
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)

# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

library(pool)
library(RSQLite)
meta_filter_v0 <- fst::read_fst("~/data/scEiaD/2021_03_17_meta_filter.fst")
# remove eric bharti 2D/3D RPE
meta_filter <- fst::read_fst("~/data/scEiaD_v2/2021_10_22_meta_filter.fst") %>% filter(study_accession != 'Bharti_Nguyen_iRPE_2D_3D')

```

# (sup?) Table 1
Studies in scEiaD
```{r}
meta_filter %>% group_by(Citation, PMID, study_accession) %>% summarise(Count = n()) %>% flextable()
```

# Figure 1 Distribution of Tissues and Cell Types in scEiaD
```{r, fig.width=8, fig.height=5}
# organ
meta_filter <- meta_filter %>% mutate(organism = factor(organism, levels = meta_filter$organism %>% unique()))
a <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Body') %>% 
  filter(Tissue != 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  # scale_y_continuous(breaks = c(2500, 5000, 7500, 10000, 12500, 84000)) +
  # ggbreak::scale_y_break(c(13000, 82000)) +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')
  
aa <- meta_filter %>% 
  filter(Tissue == 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

# eye studies
b <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = case_when(Tissue == "Choroid" ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  group_by(Organ, Tissue,organism) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) +
  xlab('')

# eye studies samples
c <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  select(Organ, Tissue,organism, study_accession) %>%
  unique() %>%
  group_by(Organ, Tissue, organism) %>% 
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0))  + xlab('')

# retina cell types
d <- meta_filter %>% filter(Organ == 'Eye') %>% 
  group_by(CellType_predict, study_accession, organism) %>%
  summarise(Count = n()) %>% filter(Count > 10) %>% ungroup() %>% 
  ggplot(aes(x=CellType_predict,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')


a_aa_b <- plot_grid(a,aa,b, ncol = 1, align = 'hv', rel_heights = c(10, 1.5,3.5), labels = 'a')
c_d <- plot_grid(c,d, rel_heights = c(2,10), ncol = 1, align= 'hv', axis = 'l',labels = c('b','c'))
plot_grid(a_aa_b, c_d, ncol = 2, rel_widths = c(1,1.2))
```

# sup fig 1 
Effect of decontx for Rho between scEiaD v2 and scEiaD v1

<!-- ```{r, fig.height=5, fig.width=3} -->
<!-- # Distribution of ML Retina Cell Types Between Mouse and Human in Pan-Retina Studies -->
<!-- Drop this figure? -->
<!-- Yes -->

<!-- pan_retina_studies <- c('SRP255195','E-MTAB-7316','SRP151023','SRP222001','SRP222958','SRP194595','SRP223254','SRP050054','SRP269634','SRP269635','SRP158528', 'SRP257758','SRP212788') -->


<!-- meta_filter %>%  -->
<!--   filter(organism %in% c('Mus musculus','Homo sapiens'),  -->
<!--          Stage == 'Mature',  -->
<!--          study_accession %in% pan_retina_studies,  -->
<!--          grepl('Amacr|Bipol|Cone|Epithel|Endoth|Horizon|Muller|Pericy|Ganglion|Rod', CellType_predict)) %>%  -->
<!--   group_by(study_accession, CellType_predict) %>%  -->
<!--   summarise(Count = n()) %>% mutate(Percentage = 100 * (Count / sum(Count)))  %>%  -->
<!--   left_join(meta_filter %>% select(study_accession, Platform, organism) %>% unique()) %>%  -->
<!--   ggplot(aes(x=organism, y=Percentage)) +  -->
<!--   geom_boxplot() +  -->
<!--   coord_flip() +  -->
<!--   ggbeeswarm::geom_quasirandom(aes(color=study_accession)) +  -->
<!--   facet_wrap(~CellType_predict, ncol = 1) +  -->
<!--   cowplot::theme_cowplot() -->
<!-- ``` -->
# Introduction
## Many published ocular atlases

## Pan Atlases contain limited retina cell types or lack integration

Tabula Sapiens, MOCA (?), BBrowser (\$\$\$), https://cells.ucsc.edu

## Data difficult to access and cross comparisons impossible

But important as seeing trends trends across different groups strongly demonstrates that 
the signal is biological, not technical. 

## What we did

Collated XX studies and processed 

# Methods

## pan mouse and pan human as non-ocular reference
Tabula Muris and 

# Results

## Updated the scEiaD backend
Our first version of the single cell Eye is a Disk (scEiaD) database contained `r meta_filter_v0$study_accession %>% unique() %>% length()` studies, three species, `r nrow(meta_filter_v0)` cells, and `r meta_filter_v0$CellType_predict %>% unique() %>% length()` curated cell types. We updated the scEiaD resource for the PLAE v0.80 web app in four important ways. First, we added two chicken retina (*Gallus gallus*) studies. Second we added an ocular outflow tract dataset. Third, we added a human pan-body reference scRNA dataset. Fourth, we used an *in silico* method to remove background gene contamination as we noticed persistent Rhodopsin expression in many non-photoreceptors cells (sup fig XX). Fourth, we increased the minimum number of detected unique transcripts per cell required for a cell to be retained from 300 to 450. 

After the dataset and quality control updates the scEiaD v02 dataset contains `r meta_filter_$study_accession %>% unique() %>% length()` studies, four species, `r nrow(meta_filter)` cells, and `r meta_filter$CellType_predict %>% unique() %>% length()` curated cell types (Fig 1).

# Comparison of PLAE to "pan" web apps with retina data
```{r}
feature_table <- cbind(c('Spectacle', 'No' ,'No', ),
                       c('UCSC Cell Browser', 'Yes', 'No', 'https://cells.ucsc.edu' ),
                       c('Tabula Sapiens', 'No', 'N/A', 'https://tabula-sapiens-portal.ds.czbiohub.org'),
                       c('PLAE','Yes','Yes', 'plae.nei.nih.gov'),
                       c('PanglaoDB', 'Yes', 'No'),
                       c('Protein Atlas', 'Yes','No', 'https://www.proteinatlas.org/humanproteome/celltype'))
colnames(feature_table) <- c('Resource Name', 'Independent Datasets', 'Harmonization', 'URL')
```

## PLAE contains over 10 fold more cells and studies than the closest comparable web app

The Spectacle web app also provides a web app to access 87,305 single cell transcriptomes across four publications.

## Extensive optimization makes the PLAE web app highly responsive despite containing huge amounts of data

## Rich functionality 

## Web accessible

## 

# Conclusion