---
title: 'PLAE web app enables powerful searching and multiple visualizations across one million unified single-cell ocular transcriptomes'
author:
  - Vinay Swamy:
      institute: bg
  - Zachary Batz:
      institute: nnrl
  - David McGaughey:
      institute:
        - bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, National Institutes of Health
  - nnrl: Neurobiology, Neurodegeneration & Repair Laboratory, National Eye Institute, National Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health

date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  word_document:
    reference_docx: word-styles-reference-01.docx
    fig_caption: yes
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
bibliography: zotero_plaeApp.bib
csl: investigative-ophthalmology-and-visual-science.csl
abstract: "PURPOSE: To create a high performance reactive web application to query single cell gene expression data across cell type, species, study, and other factors. METHODS: We updated the content and structure of the underlying data (single cell Eye in a Disk, scEiaD) and wrote the web application PLAE (https://plae.nei.nih.gov) to visualize and explore it. RESULTS: The new portal provides quick visualization of over a million individual cells from vertebrate eye-and body transcriptomes encomopassing four species, 60 cell types, six ocular tissues, and 23 body tissues across 37 publications. To demonstrate the value of this unified pan-eye dataset, we replicate known neurogenic and cone macula markers as well as propose six???? new cone human region markers. CONCLUSION: The PLAE web application provides the eye community a powerful and quick means to test hypotheses related to gene expression across a highly diverse, community-derived database."

keywords: "RNA-seq, lens, cornea, retina, RPE, Snakemake, web, app"
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)
library(glue)
library(tictoc)
#tic()()
data_dir_vGiga <- '/Volumes/McGaughey_S/data/scEiaD/'
data_dir_vPLAE <- '/Volumes/McGaughey_S/data/scEiaD_2022_02/'
#data_dir_vGiga <- '~/data/scEiaD_gigascience/'
#data_dir_vPLAE <- '~/data/scEiaD_2022_02/'
# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

library(pool)
library(RSQLite)
meta_filter_v0 <- fst::read_fst(glue("{data_dir_vGiga}/paper_data/2021_03_17_meta_filter.fst")) %>% 
  
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                             TRUE ~ CellType))

meta_filter <- fst::read_fst(glue("{data_dir_vPLAE}/meta_filter.fst")) %>% filter(study_accession != 'Bharti_Nguyen_iRPE_2D_3D') %>% 
  
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                             TRUE ~ CellType),  
         Tissue = case_when(grepl('Iris', Tissue) ~ 'Iris',
                            TRUE ~ Tissue),
         Tissue = case_when(grepl('^RPE|^Choroid', Tissue) ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Citation = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ 'Swamy VS, Fufa TD, Hufnagel RB, McGaughey DM Building the Mega Single Cell Transcriptome Ocular Meta-Atlas Gigascience 2021 Oct 13;10(10)',
                              TRUE ~ Citation),
         PMID = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ '34651173',
                          TRUE ~ PMID),
         study_accession = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ 'SRP329495',
                                     TRUE ~ study_accession))


load('data/diff_data_pull.Rdata') # from src/diff_data_pull.R

data_dir_vGiga <- '/Volumes/McGaughey_S/data/scEiaD/'
data_dir_vPLAE <- '/Volumes/McGaughey_S/data/scEiaD_2022_02/'
scEiaD_2020_v00 <- pool::dbPool(RSQLite::SQLite(), dbname =  glue('{data_dir_vGiga}/MOARTABLES__anthology_limmaFALSE___5000-transform-counts-universe-batch-scVIprojectionSO-8-0.1-50-5.sqlite'))
scEiaD_2020_v01 <- pool::dbPool(RSQLite::SQLite(), dbname =  glue('{data_dir_vPLAE}/MOARTABLES__anthology_limmaFALSE___4000-counts-universe-study_accession-scANVIprojection-15-5-0.1-50-20__pointRelease01.sqlite'))

# Celltypes in celltype but not celltype_predict
dropped <- meta_filter$CellType[!meta_filter$CellType %in% meta_filter$CellType_predict]

#toc()

num1 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num2 <- meta_filter %>% filter(Organ != 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num3 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(CellType) %>% unique() %>% length()
```


# Introduction

The retina is composed of six major cell populations: the photoreceptors, horizontal, bipolar, amacrine, ganglion, and non-neuronal cells [@masland_cell_2011]. These six populations can be further divided into dozens of cell types [@yan_cell_2020]. Furthermore, human disease of the eye can derive from damage to specific cell types. Glaucoma is characterized by damage to the retinal ganglion cells while cone and cone-rod dystrophies are characterized by photoreceptor degeneration. Efforts to untangle the transcriptomics of these diverse cell types with single cell transcriptomics date to the early 2000s when researchers [@hagstromConePigmentGene2000b; @jabsEvidenceP2XP2X2000; @liLocalizationDopamineD1receptor2004; @lindqvistSingleCellRTPCR2002a; @reidRetinoschisinPhotoreceptorsecretedProtein2003]  used RT-PCR approaches to amplify select transcripts in small numbers of cells. Later, researchers used microarray platforms to analyze a broader array of transcripts[@cherryDevelopmentDiversificationRetinal2009; @gustincichGeneDiscoveryGenetically2004a; @kimIdentificationMolecularMarkers2008a; @roeschTranscriptomeRetinalMuller2008; @trimarchiIndividualRetinalProgenitor2008; @trimarchiMolecularHeterogeneityDeveloping2007b; @wahlinMethodAnalysisGene2004a]

The modern era of single cell transcriptomics ha been marked by the introduction of droplet-based technologies, which enabled quantitation of thousands of cells in a single experiment. This approach was pioneered by Macoksko et al. and enabled the profiling of gene expression in 44,808 single cells in the adult mouse retina. [@kleinDropletBarcodingSingleCell2015; @macoskoHighlyParallelGenomewide2015a]. The commercialization of the droplet approach has enabled broad access to this technology and the ocular community have responded by publishing several dozen ocular studies using single cell transcriptome technology since Macosko publication. 

These publications's foci includes, but are not limited to, studying developmental transcriptomic dynamics, identifying genes to distinguish core cell types, proposing finer gradations of cell types within existing categories, and comparing ocular transcriptomes between organisms. Independent querying of these datasets requires a laborious series of steps. 

Unfortunately manuscripts can only present a sliver of the underlying data and further exploration is laborious, technically challenging, and computationally expensive. Briefly, one would have to identify the relevant publication, find the associated data deposit, download the cell - count matrix, find the cell label table (and optionally the t-SNE or UMAP coordinates), and then load the data into R or python to run the desired queries. If one desired to compare across publications, then one would also have to align the cell type labels and should re-create the count quantification with consistent bioinformatic tooling. It is desirable, perhaps even necessary, to use multiple datasets to assess gene expression to confirm whether a gene expression pattern is consistent and reproducible. Reliance on a single dataset can be problematic as there can be technical issues - for example the large consortium dataset GTEx has persistent contamination from highly expressed genes across tissues [@nieuwenhuisConsistentRNASequencing2020]. 

Web-based resources to query gene expression can democratize access to large datasets as the computationally intensive steps above can be run once and then shared to anyone with a internet connected device. Single cell RNA-seq web resources that contain ocular tissues include the UCSC Cell Browser, Tabula Sapiens, PanglaoDB, and the Protein Atlas [@consortiumTabulaSapiensSingle2021; @franzenPanglaoDBWebServer2019; @karlssonSingleCellType2021; @speirUCSCCellBrowser2021]. Primary limitations to these resources include incomplete cell type labels and sample metadata that make ocular focused queries challenging. In contrast to these, the Spectacle resource is an ocular specific web app which has curated over a million cells from dozens of independent resources [@voigtSpectacleInteractiveResource2020]. However each data deposit is independently processed, which makes cross study queries impossible. We recently created a unified ocular data resource which placed dozens of independent single-cell RNA-seq ocular and non-ocular datasets in a single combined space [@swamyBuildingMegaSinglecell2021]. 

To broaden the usage of this powerful meta-atlas, we restructured the data into a custom sqlite-based database and built a reactive web application to access it. As many new studies were published since our last data build, we re-ran the scEiaD pipeline in March of 2022. In aggregate we have collated `r meta_filter$study_accession %>% unique() %>% length()` datasets across `r meta_filter$PMID %>% unique() %>% length()` publications, four species, and `r meta_filter$Tissue %>% unique() %>% length()` tissues. We hand-curated `r meta_filter$CellType_predict %>% unique() %>% length()` cell type labels, totaling `r meta_filter %>% filter(!is.na(CellType)) %>% nrow()` cells and used those ground-truth labels as the basis for a machine learning-based algorithm which labels the `r length(meta_filter$CellType_predict[!is.na(meta_filter$CellType_predict)]) - length(meta_filter$CellType[!is.na(meta_filter$CellType)])` more cells. Quality control metrics were tuned and applied to remove lower quality cells. In the end, the single cell Eye in a Disk (scEiaD) database contains `r meta_filter %>% nrow()` cells and 1,509,179,347 non-zero gene/cell expression values. We demonstrate how this unified resource can be used to identify genes that consistently distinguish cell types across studies and species, find genes which have species-specific expression cell type patterning, demonstrate how our pre-computed differential testing can identify markers between similar progenitor cell types, and finally provide a R-based analysis document which end-to-end demonstrates how we use scEiaD to run a custom differential gene expression test to validate and propose new genes that distinguish macula from peripheral cones in humans.

# Methods

## scEiaD pipeline upgraded and the database re-built

The scEiaD pipeline is largely the same as in Swamy et al. [@swamyBuildingMegaSinglecell2021]. Very briefly, we collected the raw fastq files and quantified with the kallisto / bustools system [@melstedBarcodeUMISet2019; @melstedModularEfficientPreprocessing2019]. The quantified counts are merged into one matrix that is used for the downstream integration with scvitools create a batch corrected lower dimensional space. The raw counts and scANVI batch corrected lower dimensional space is sent into Seurat for quality control, clustering, and 2D UMAP visualization. We use our previously published xgboost-based machine learning approach to transfer cell type labels to unlabelled cells [@swamyBuildingMegaSinglecell2021].  

The following alterations were made: First, the cutoff to retain a cell was raised from 200 unique genes quantified to 300. Second, we use the DecontX algorithm[@yangDecontaminationAmbientRNA2020] from the celda R package (version 1.9.2) to automatically remove ambient RNA contamination on a per-study basis. Third, we altered the procedure for aligning gene names across species (see below for details). Fourth, we use the scANVI (scvitools based) batch correction method, which leverages the known cell type labels in the batch correction process [@xuProbabilisticHarmonizationAnnotation2021].

A custom cutoff for the minimum required number of unique genes identified per cell were used for the SRP362101 cornea dataset as an abnormally high number of cells were returned from our default value of 300. We set the cutoff for this studies to be 800 as this value resulted in approximately the same number of cells being returned as the authors reported in their papers. 

We found that while nearly all datasets we curated could be integrated as a whole, a small number had strange behavior in the two dimensional UMAP space and/or nearly all cells from a study were machine labelled as a single cell type. These studies were hand-removed from the resource. 

## Cross species gene alignment

As we have four species in scEiaD (*Gallus gallus*, *Macaca fascicularis*, *Mus musculus*, and *Homo sapiens*) that we unify, we must identify shared genes. We found homologs/orthologs by pulling from the BioMart database "Ensembl Genes 105" and using *Homo sapiens* as the reference. We then selected orthologous gene names from *Gallus gallus*, *Macaca fascicularis*, and *Mus musculus*. This generates a table linked by Ensembl gene IDs and gene names. We then use a full join on non-duplicated Ensembl IDs between mouse and human. As we noticed a small number of gene names failed to be linked in this manner we ran another join on the remaining non-aligned genes, using gene name. This procedure gave us 17,769 shared human and mouse genes. To align the other species we again joined on Ensembl gene ID, removing genes where one chicken or macaque gene aligned to multiple human/mouse genes. In cases where multiple chicken or macaque genes aligned to one human/mouse gene, then we aggregated the counts by sum. In cases where there was no corresponding chicken or macaque gene, we filled in zeros. After these steps we had, as expected, 17,769 genes. 

## Metadata curation

Every study brought into scEiaD was hand curated to identify unique biological samples, published paper ID (where available), organ, tissue (e.g. Cornea), source (iPSC based or tissue), single cell platform (e.g. 10Xv2, DropSeq, etc.). We also, where possible, annotated retina region (e.g. macula), sex, and age or developmental stage. Samples with perturbations (genetic or treatment) were not included. To import individual cell type labels (e.g. Rod, Muller Glia), we wrote custom code for each study that made this table available to link their cell type assignments back to the matching barcoded cell. We curated two types of cell type labels: "CellType" and "SubCellType" where the former is our normalized cell types taken from the published labels. We normalized names for CellType, for example changing "MG" to "Muller Glia" and also dropping more detailed cell type assignments like off and on bipolar cells for Bipolar Cells (as few studies went into this level). We did retain the original published (except for the broader name normalization) labels under SubCellType. 

## Differential gene testing

The aggregateAcrossCells function from the scuttle R package was used to aggregate gene counts across species, study accession, and category where category was either Cluster, CellType, or CellType (Predict) assignments for a cell [@mccarthyScaterPreprocessingQuality2017]. If there were fewer than 50 cells in organism - study accession - category combination, they were discarded for the differential test. Any genes with a sum of 0 counts after aggregation were removed. The aggregated matrix was imported into DESeq2 using the DESeqDataSetFromMatrix function with the design given as "~study_accession + category". Contrasts were extracted either with the target category (e.g. "Cone") against all remaining or in a pair wise manner ("Cone" vs "Rod") with the DESeq2 "results" function [@loveModeratedEstimationFold2014a]. 
<!-- Genes were considered to be significantly differentially expressed if they have a padj less than $1 x 10^-5 and a abs(log2FoldChange) > 2.  -->

## scEiaD structure and web app optimization

The web app at plae.nei.nih.gov can display gene expression across over a million cells in only a few seconds. This speed is only possible due to custom data structures that were optimized for data query efficiency. Like eyeIntegration's EiaD database [@swamyEyeDiskEyeIntegration2019a], scEiaD is a sqlite database with tables for the core categories. This allows the app to initialize in about 30 seconds on a cloud server with memory usage under 8 GB. For the gene by cell expression matrix, the data was transformed into a "long" format where there are three columns: gene, cell barcode, and gene count value. This allows for cell - gene level queries to be completed, again with minimal memory usage, in under 0.5 seconds. To generate the aggregated information (in the Dot Plot, Expression Plot, and tables) we pre-calculated all queries by running a dplyr "group by" operation on all column fields. On user given query, the data is further aggregated to the user request. This allows for complicated queries to complete in seconds rather than minutes. To compare web app loading times for useful tasks we queried https://plae.nei.nih.gov, http://singlecell-eye.org, and http://tabula-sapiens-portal.ds.czbiohub.org, on 2022-08-24. For Spectacle, we used the "Single-Cell RNA-Seq Analysis of Retinal Development..." 10X dataset from Clark et al. for all timings. Timings are accurate to the second. If the functionality tested was not available for a certain web app, it was represented by a blank space. 

## Data reproducibility

The code base for scEiaD is available at github.com/davemcg/scEiaD; the commit corresponding to this manuscript is XXXXXXX. There are three relevant Snakemake pipelines which were used to create scEiaD: SnakeQUANT, which was used to quantify the gene / cell expression from the raw fastq files, SnakePOP, which runs the multi-parameter integration methods, and SnakeSCEIAD, which uses the optimal parameters identified in SnakePOP to run the cell type machine learning and differential gene testing that is incorporated into the scEiaD sqlite database for the app. The code base for the web app (along with local installation instructions) is available at github.com/davemcg/scEiaD/plaeApp; the commit corresponding to this manuscript is XXXXXXXXXXXX.

# Results

## New studies and improvements to the scEiaD database

Our first version of the single cell Eye is a Disk (scEiaD) database contained `r meta_filter_v0$study_accession %>% unique() %>% length()` studies, three species, `r nrow(meta_filter_v0)` cells, and `r meta_filter_v0$CellType_predict %>% unique() %>% length()` curated cell types [@swamyBuildingMegaSinglecell2021]. We updated the scEiaD database for the PLAE v0.91 web app in six important ways. First, we added a new species, chicken (*Gallus gallus*) [@yamagataCellAtlasChick2021]. Second we added two ocular outflow tract datasets, one brain choroid plexus, and three cornea datasets to enhance coverage across the eye [@collinSingleCellAtlas2021; @gautamMultispeciesSinglecellTranscriptomic2021; @daniCellularSpatialMap2021; @ligockiMolecularCharacteristicsSpatial2021; @patelMolecularTaxonomyHuman2020; @vanzylCellAtlasAqueous2020]. Third, we added a human pan-body reference scRNA dataset to allow for non-ocular comparisons [@heSinglecellTranscriptomeProfiling2020]. Fourth, we used an *in silico* method to remove background gene contamination as we noticed persistent Rhodopsin expression in many non-photoreceptors cells (`r supFig_cap(name = 'decontX', display = 'cite')`). Fifth, we increased the cell retention cutoff for detected unique transcripts per cell from 200 to 300. Sixth, we carried over and harmonized several common cell type labels from the Tabula Muris project to more easily provide non eye cell type comparisons [@SinglecellTranscriptomics20]. As before, the pipeline's scVI's integration parameters were chosen rigorously by testing inegration performance across a wide range of latent dimensions and number of highly variable genes in our scPOP framework [@swamyBuildingMegaSinglecell2021]. Each study in the resource were hand assessed to determine whether they had any unusual properties in regards to machine learnt cell type proportions or position in the 2D UMAP space. The small number of outlier studies were removed from the resource. 

After the dataset and quality control updates the scEiaD v2022-03-21 dataset now contains `r meta_filter$study_accession %>% unique() %>% length()` studies, four species, `r nrow(meta_filter)` cells, and `r meta_filter$CellType_predict %>% unique() %>% length()` curated cell types (`r fig_cap(name = 'sup_scEiaD_data_overview', display = 'cite')`). The scEiaD database contains `r num2` non-ocular human and mouse tissues and `r num1` ocular tissues across human, chicken, macaque, and mouse (`r fig_cap(name = 'scEiaD_data_overview', display = 'cite')`). This resource is only possible due to the publication of `r meta_filter$PMID %>% unique %>% length()` single cell resources that scEiaD draws from (`r supTab_cap(name = 'studies_table', display = 'cite')`). To recognize these papers, we prominently feature each of these studies along with a direct link to their citation, where possible, on the loading page of plae.nei.nih.gov.

```{r, fig.height=14, fig.width=12, fig.cap = scEiaD_data_overview_cap}


scEiaD_data_overview_cap <- fig_cap(name = 'scEiaD_data_overview', caption = glue::glue("Distribution of tissues and cell types in scEiaD. A. Number of cells for each ocular cell type. B. Number of studies for each ocular predicted cell type (over 10 cells), colored by species. C. Tabulation of the number of studies present across {num1} ocular and {num2} non-ocular tissues, D. Web app timings (lower is better) for important functions compared between PLAE, Spectacle, and Tabula Sapiens (see methods for more details)."))

study_count_by_tissue <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  #filter(Organ == 'Eye') %>% 
  select(Organ, Tissue,organism, study_accession) %>%
  unique() %>%
  group_by(Organ, Tissue, organism) %>% 
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(Organ = factor(Organ, levels = c('Eye','Body'))) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, label = Count, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position='stack') +
  shadowtext::geom_shadowtext(position = position_stack(vjust = 0.5)) +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0))  + xlab('') +
  ggforce::facet_col(~Organ, scales = 'free_y', space = 'free')

# retina cell types
retina_ct <- meta_filter %>% filter(Organ == 'Eye') %>% 
  pull(CellType) %>% unique()

cell_counts <- meta_filter %>% filter(Organ == 'Eye', CellType_predict %in% retina_ct) %>% 
  mutate(CellType_predict = case_when(is.na(CellType_predict) ~ 'Unlabelled', TRUE ~ CellType_predict)) %>% 
  group_by(CellType_predict, organism) %>%
  summarise(Count = n()) %>% 
  mutate(sCount = log2(Count)) %>% 
  ggplot(aes(x=CellType_predict,y=sCount, label = Count, fill = organism, group = organism)) +
  ylab('log2(Number of Cells)') +
  geom_bar(stat='identity', position='stack') +
  shadowtext::geom_shadowtext(position = position_stack(vjust = 0.5)) + 
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')

study_counts_by_cell <- meta_filter %>% filter(Organ == 'Eye', CellType_predict %in% retina_ct) %>% 
  mutate(CellType_predict = case_when(is.na(CellType_predict) ~ 'Unlabelled', TRUE ~ CellType_predict)) %>% 
  group_by(organism, CellType_predict, study_accession) %>%
  summarise(Count = n()) %>% 
  filter(Count > 10) %>% 
  summarise(Count = n()) %>% 
  mutate(Count = case_when(CellType_predict != 'Unlabelled' ~ Count)) %>% 
  mutate(sCount = log2(Count)) %>% 
  ggplot(aes(x=CellType_predict,y=Count, label = Count, fill = organism, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position='stack') +
  shadowtext::geom_shadowtext(position = position_stack(vjust = 0.5)) + 
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')

timings <- rbind(
  c(1,7,1),
  c(9,13,8),
  c(3,NA,2),
  c(2,NA,NA),
  c(2,1,NA),
  c(2,NA,NA),
  c(2,1,NA),
  c(4,41,NA)
) %>% data.frame()

row.names(timings) <- c('Initial Page',
                        'Cell Type Colored UMAP',
                        'Abca4 Expression UMAP',
                        'Boxplot of Abca4 Expression By Cell Type Across All Studies',
                        'Boxplot of Abca4 Expression By Cell Type Across One Study',
                        'Dotplot of Abca4, Pax6, Nrl, and Atoh7',
                        'Heatmap of Abca4, Pax6, Nrl, and Atoh7',
                        'Table of Differentially Expressed Genes between Bipolar and Cone Cells'
                        
) %>% str_wrap(., width = 32)

colnames(timings) <- c('PLAE',"Spectacle",'Tabula Sapiens\n(CELLxGENE)')

breaks_fun <- function(x) {
  if (max(x) > 30) {
    seq(0,40,20)
  } else if (max(x) > 9) {
    seq(0, round(max(x)), 5)
  } else if (max(x) >= 6) {
    seq(0, round(max(x)), 2)
  } else if (max(x) >= 3) {
    seq(0, round(max(x)), 1)
  } else {
    seq(0,2,1)
  }
}

timing_plot <- timings %>% as_tibble(rownames = 'Test') %>% 
  pivot_longer(-Test) %>% 
  mutate(Test = factor(Test, levels = row.names(timings))) %>% 
  mutate(Text = case_when(is.na(value) ~ 'Not possible',
                          TRUE ~ '')) %>% 
  ggplot(aes(x=name,y=value, label = Text)) + 
  geom_bar(stat= 'identity') + 
  geom_text(aes(y=0), hjust = 0, angle = 90) +
  facet_wrap(~Test, scales = 'free_y', ncol = 2) + 
  xlab('Resource') + 
  ylab('Seconds to Load') + 
  cowplot::theme_cowplot() +
  scale_y_continuous(breaks = breaks_fun)





ab <- cowplot::plot_grid(cell_counts + theme(legend.position = "none", axis.ticks.y = element_blank()), 
                         study_counts_by_cell + theme(axis.ticks.y = element_blank()), 
                         nrow = 1, 
                         rel_widths = c(1,1), labels = c('a','b'))

cd <- cowplot::plot_grid(study_count_by_tissue + theme(axis.ticks.y = element_blank()), 
                         timing_plot, 
                         nrow = 1, 
                         rel_widths = c(0.7,1),
                         labels = c('c','d'))

cowplot::plot_grid(ab, cd, ncol = 1, rel_heights = c(1,0.8))

```


## PLAE resource contains several crucial features relative to other single cell resources

Several single cell RNA-seq web-based resources contain ocular data (`r tab_cap(name = 'web_resources', display = 'cite')`). The UCSC Cell Browser, Tabula Sapiens portal, PanglaoDB, and Protein Atlas are not ocular focused, but do contain some ocular data. The PanglaoDB and UCSC are resources which have ingressed a large number of published single cell datasets The UCSC Cell Browser has around sixty thousand ocular single cell transcriptomes while the PangloaDB has about forty-seven thousand. However, these resources provide no curation of the metadata and process the data for each study individually. Thus each study is effectively silo-ed off from the rest and cross comparisons are impractical. The Tabula Sapiens project is an effort to sequence all cell type transcriptomes in humans; while it only contains around around ten thousand ocular cells it can be more readily cross compared to non-ocular tissues / cells. A more substantial downside is that the cell type labels can be insufficient - for example rods and cones are not independently identified but rather labelled as photoreceptors. The Protein Atlas resource has recently used only one retina single cell RNA-seq paper to provide cell type label focused information for gene queries (e.g. one can see whether a gene of interest is enriched in Cones). This is problematic as it is unclear whether signal found is specific to the one dataset they use or is common across independent species and datasets.

The Spectacle web resource is the closest comparison to PLAE, as it is a ocular focused resource which has collated a large number of ocular single cell RNA-seq resources (both organoid and tissue based datasets total well over a million cells) and made them available on a web app. PLAE is distinguished from these resources by hand curation of the cell and sample level metadata, re-processing of the data from the raw sequence in a consistent manner, and, crucially, full availability of the underlying data for outside usage and integrating the independent studies to deliver cross study comparisons. 


```{r web_resources, fig.height = 15, fig.width = 20, echo = F, message = F, results = 'asis', fig.cap=web_resources_cap}
#tic()()
i = 16.5 # width of the side borders in the word_document output (in centimeters)
w = i*0.3937 # width of the side borders in the word_document output (in inches)



web_resources_cap <- tab_cap(name = 'web_resources', caption = 'Comparison of PLAE with other web resources that contain single cell ocular datasets, as of 2021-03-15. The Protein Atlas is asterisked as while it contains independent datasets, it only has one ocular study. ')

cat("<div id=\"tbl:web_resources\">")

feature_table <- rbind(
  c('PLAE','Yes','Yes', 'Yes', '1136 ', 'https://plae.nei.nih.gov'),
  c('Spectacle', 'Yes' ,'No', 'No', '1500', 'https://singlecell-eye.org'),
  c('UCSC Cell Browser', 'Yes', 'No', 'Yes', '62', 'https://cells.ucsc.edu' ),
  c('Tabula Sapiens', 'No', 'N/A', 'Yes', '10', 'https://tabula-sapiens-portal.ds.czbiohub.org'),
  c('PanglaoDB', 'Yes', 'No', 'Yes', '47', 'https://panglaodb.se'),
  c('Protein Atlas', 'Yes*','No', 'Yes', '5',  'https://www.proteinatlas.org/humanproteome/celltype')) %>% 
  as_tibble()
colnames(feature_table) <- c('Resource Name', 'Independent\nDatasets', 'Harmonization', 'Ocular and\nnon-Ocular\n Data', 'Ocular Cell\nCount\n(thousands)', 'URL')
feature_table %>% flextable() %>%  fontsize(size = 10, part = 'all') %>% autofit() %>% fit_to_width(7)#width(., width = c(1,0.75,0.75,0.75,0.75,3))
# as of 2021 10 25
#toc()()

#Output Caption (* * for italics)
cat(paste0("\n*",web_resources_cap,"*\n"))
# Output </div>
cat("</div>")
```

While PLAE and Spectacle (https://singlecell-eye.org) both contain large numbers of ocular-related scRNA datasets, they are structured differently[@voigtSpectacleInteractiveResource2020]. PLAE is built around the scEiaD database, which use a multi-stage pipeline to build and identify a high performing scVI-based single cell RNA eye model that is used to integrate all datasets together [@swamyBuildingMegaSinglecell2021]. This allows for gene queries to be constructed and analyzed across studies. In contrast, Spectacle, as of March 2022, is a compilation of study-level datasets and thus you cannot run queries across studies. Spectacle contains no non-ocular dataset and the cell types labels have not been curated or harmonized between studies (for example different studies can use MG, Mueller, Müller Glia, or other variations to refer to the same cell type). Finally, in contrast to PLAE, Spectacle does not make their datasets available for download in any form. 

## Extensive optimization makes the PLAE web app highly responsive despite containing huge amounts of data

There are two interwoven reasons why the PLAE app is quick, despite the huge amount of data it contains. First, the data structure of scEiaD is a sqlite database with pre-calculated tests stored in tables. A sqlite database and its indexing functionality allows large amounts of data to be stored for quick retrieval with minimal loading time and memory usage. The major downside is that this approach requires a large amount of disk storage, which can make local (on personal computer) usage difficult; the scEiaD v2022-03-22 sqlite database is about 368 GB. Second, the visualization code in PLAE (e.g. the UMAP view) pull from the custom database directly and use a highly optimized plotting system (scattermore) to reduce drawing time. In contrast, starting from the loading page of Spectacle, it takes approximately eleven seconds to plot the cell type metadata for a several thousand cells across across a single study (tested on 2022-08-26). Within the same time, PLAE can plot the cell type labels for over a million cells across `r meta_filter_v0$study_accession %>% unique() %>% length()` studies (`r fig_cap(name = 'sup_scEiaD_data_overview', display = 'cite')d`). The speed of the PLAE app is also comparable to the CELLxGENE browser serving the Tabula Sapiens resource while providing additional functionality. 

## Rich web-available visualizations 

PLAE has several ways to visualize the data: one being the 2D UMAP view which, side-by-side, shows gene - cell expression and metadata - cell values. So thus a user can see Crx expression across over a million cells and separately see cell type labels for those same million cells. The 2D UMAP view also has a zoom function, so a user can focus on a subset of cells. Clicking on the plot will show metadata for the five nearest cells to the click, which is useful to seeing what kind of cells are present in a small area (`r fig_cap(name = 'PLAE_screenshots', display = 'cite')`a). 

For users interested in seeing how expression of a gene relates to study or other covariates, the Expression Plot view displays expression by cell type, predicted cell type, or cluster. The values can then be split and colored by multiple covariates like study or organism. This allows a user to see whether a gene pattern is consistent across studies or species. We also offer a Dot Plot view which can display a large number of genes in a space efficient manner. This allows a users to see whether a set of genes has related (or not) expression patterns across multiple cell types (`r fig_cap(name = 'PLAE_screenshots', display = 'cite')`b and c). 

While scEiaD does not have spatial scRNA seq datasets as of 2022, we do leverage existing knowledge of the structure of the retina to show *in silico* visualization of genes by cell type / layer in a cartoon cross section of the retina. The structure of the cartoon retina was based on the human retina. This unique visualization displays the major cell types of the retina and the RPE / choroid layer behind it. The cell types are colored by the relative amounts of expression in the user selected gene. Like all visualizations in PLAE, the user can use powerful filtering options to only display data from certain metadata characteristics, like species or publication (`r fig_cap(name = 'PLAE_screenshots', display = 'cite')`d).

We have also added a Heatmap visualization which summaries the relative expression differences between user selected cell types or clusters across human, macaque, and mouse. This plot differs from the Dot Plot view as the expression values are relative instead of absolute. That is to say that while the Dot Plot view is proportionate to counts, the Heatmap view is showing the difference in expression between a given cell type (or cluster) and all other cells. This view is potentially useful if a user is especially interested in gene expression patterns across a variety of cell types. 

```{r PLAE_screenshots, fig.height = 18, fig.width = 10, echo = F, message = F, fig.cap=PLAE_screenshots_cap}
PLAE_screenshots_cap <- fig_cap(name='PLAE_screenshots', caption = 'Screenshots from PLAE app. A. UMAP view showing CRX expression with cell type labels, B. Information table showing expression of Crx across celltypes and organism. C. Box plot of selected genes in a user selected cell types across different studies. D. In silico in-situ of Rho. E. Dotplot demonstating efficient display of gene expression across many cell types.')
knitr::include_graphics('fig/fig1.png')
```

<!-- ## Leveraging the high diversity of studies to identify thousands of consistently differentially expressed genes across ocular cell types -->



<!-- Across the cell types of the eye, we find `r diff_tab %>% nrow()` gene - celltype combinations that are significantly differentially expressed. We define significant as a log2 fold change greater than two as well as a p-value (corrected) less than $1 x 10^-5$. We find a mean of  -->
<!-- `r diff_tab_sup$"Homo sapiens" %>% mean() %>% round()`,  -->
<!-- `r diff_tab_sup$"Mus musculus" %>% mean() %>% round()`, and  -->
<!-- `r diff_tab_sup$"Macaca fascicularis" %>% mean() %>% round()` differentially expressed genes in human, macaque, and mouse respectively (`r supTab_cap(name = 'diff_tab_sup_tab', display = 'cite')`). These variable results directly reflect the differing number of independent studies - mouse and human have substantially more than macaque (`r supTab_cap(name = 'well_supported_celltypes_tab', display = 'cite')`). The significant filtered differential table is made available as supplementary file "differential_expression_table.csv.gz" and the full differential tables are available for download at plae.nei.nih.gov -> Data.  -->

## Well supported ocular cell types identify high confidence cell type markers

We used a "pseudo-bulk" based approach where the counts for a category (CellType, Cluster, or CellType (Predict)) were summed for each category - organism - study grouping (see Methods). This created a counts matrix which has statistical properties approximating "bulk" RNA seq experiment matrices. This allowed the use of the more mature bulk RNA-seq tooling. This approach has also been found to be more specific in identifying differentially expressed genes in single cell RNA-seq experiments[@squairConfrontingFalseDiscoveries2021]. Once the pseudo-bulk matrix was constructed, we use the DESeq2 differential gene expression analysis software to identify differentially expressed genes. We tested each species (Human, Macaque, Mouse) separately. Chicken was not included as we had an insufficient number of independent studies to conduct differential testing. The model used for DESeq2 was "~study_accession + category" in order to regress out study specific differential gene expression. After running these tests we extracted differentially expressed genes with two types of contrasts: either against all remaining categories (e.g. cones versus all remaining cell types) or in a pair-wise manner (cone against rod). Unless stated otherwise, we will be discussing the results of the "against all" contrast for the remainder of the paper. 

Cell type marker genes can be proposed on the basis of existing knowledge about their function and on high expression differences when comparing to other cell types. We can further narrow down the list of markers by using the high diversity of studies and organisms in scEiaD to propose a set of community supported cell type markers. We first identify a set of well supported cell types which we define as cell types which have two or more independent studies across both human and mouse (`r supTab_cap(name = 'well_supported_celltypes_tab', display = 'cite')`). These cell types were then assessed to identify differentially expressed genes which met the following criteria across human, mouse, and macaque: 1. padj less than $1 x 10^-4$
in two or more species, 2. mean log2 fold change greater than 2, and 3. mean padj less than $1 x 10^-5$. A final filter was applied to remove genes as candidates if they were differentially expressed (using the criteria above) in more than three different cell types. This left us with `r top_genes %>% nrow()` gene - cell type markers, a substantial reduction from the `r diff_tab %>% nrow()` we had initially. We provide a table of these genes as supplementary file "consistently_differentially_expressed_genes.csv.gz." 

To visualize these top markers we select up to eight genes (ordered by the mean log2 fold change across organisms) per well supported cell type for each and plot the log2 fold change of the gene expression relative to all other cell types (`r fig_cap(name = 'consist_genes_heatmap', display = 'cite')`). We see that human, mouse, and macaque have similar expression patterning of the proposed markers across the well supported cell types. The cell types are arranged by the columns by how related the expression patterns are. We see how the differentiated cell types of the retina are grouped together with the exception of the Müller glia, which are near the RPC from which they derive from. The photoreceptors are also next to each other.

To assess whether this marker selection was identifying known or potentially novel marker genes we ran several sets of queries against PubMed. We searched for the gene name or the gene name plus " AND Retina" across all well supported genes against the PubMed database, counting how many citations were returned. In contrast we also selected 1000 random genes that were not considered differentially expressed and ran the same PubMed searches. We find a substantial enrichment of citations for our well supported differentially expressed genes (t test, p < 6.7x10-98 and p < 2.4x10-12) for the gene and gene " and retina" PubMed searches when compared against the 1000 random genes, respectively. To display the results of this search for each individual gene we show visually whether the number of citations returned is less or equal than the median citation count for the random gene set ("few"), more than the median ("some"), or more than the mean ("many"). We see that nearly all genes in (`r fig_cap(name = 'consist_genes_heatmap', display = 'cite')`) reflect existing knowledge. 

Genes which have little existing knowledge can be considered 

Of course other possibilities for ranking genes are possible. We show in `r supFig_cap(name = 'PLAE_screenshots_supplemental', display = 'cite')`b and c how you can directly use our powerful web app to select candidate bipolar cell differentially expressed genes from the differential testing table and then quickly display the expression pattern of those genes in a compact heatmap visualization. The wealth of visualization options in plae allow for quick prototyping and hypothesis testing of gene expression patterning across a large number of cells and studies. 

```{r echo = F, message = F, }

# tic()
# ws_oct <- ws_table %>% filter(`Well Supported` == 'Yes') %>% pull(`CellType (Predict)`)
# 
# consist_diff <- scEiaD_2020_v01 %>%
#   tbl("diff_testing") %>%
#   filter(Base %in% ws_oct) %>%
#   filter(Group == 'CellType (Predict)',
#          Against == 'All',
#          Organism %in% c('Homo sapiens','Mus musculus')) %>%
#   group_by(Base, Gene) %>%
#   summarise(`mean log2FC` = mean(log2FoldChange), `sig count` = sum(padj < 1e-5), `mean padj` = mean(padj)) %>%
#   filter(`mean log2FC` > 2.463, `sig count` > 1, `mean padj` < 1e-5) %>%
#   collect() %>%
#   mutate(GENEID = str_extract(Gene, 'ENS\\w+')) %>%
#   left_join(gene_tab, by = 'GENEID') %>%
#   relocate(SYMBOL, GENEID) %>%
#   dplyr::rename(Symbol = SYMBOL, ID = GENEID) %>%
#   arrange(-`mean log2FC`)

# # id diff genes only seen once across the cell types
# uniq_g <- consist_diff %>% group_by(Symbol) %>% count() %>% filter(n == 1)
#consist_diff %>% group_by(Base) %>% summarise(Count = n())
#consist_diff %>% filter(Symbol %in% uniq_g$Symbol) %>% group_by(Base) %>% summarise(Count = n())

sup_tab_consist_diff <- top_genes %>% select(Symbol, Base) %>% left_join(consist_diff, by = c('Symbol','Base')) %>% group_by(Base) %>% summarise(Count = n()) %>% ggplot(aes(x=`Base`, y = `Count` )) + geom_bar(stat='identity', position = 'dodge') + cowplot::theme_cowplot() + coord_flip() + xlab("CellType (Predict)")
#toc()
```

```{r consist_genes_heatmap, fig.height=22, fig.width=10, echo = F, message = F,  fig.cap=consist_genes_heatmap_cap}
consist_genes_heatmap_cap <- fig_cap(name='consist_genes_heatmap', caption = 'Rows are genes, columns are cell types and organisms. Rows are split by what cell type the gene is a marker for. Columns are split by the cell types. More red is a higher log2 fold change of the gene - cell type relative to all other cell types and blue is more negative. Relative number of citations are shown for the gene or gene " and retina" when searching against PubMed compared to 1000 randomly chosen genes.')


load('data/top_marker_pmid.Rdata')

# tic()
# g <- consist_diff %>% filter(Symbol %in% uniq_g$Symbol) %>% group_by(Base) %>% slice_max(n=8, order_by = `mean log2FC`) %>% mutate(g = paste0(Symbol, ' (', ID, ')')) %>% pull(g)
# 
# consist_diff_long <- scEiaD_2020_v01 %>%
#   tbl("diff_testing") %>%
#   filter(Gene %in% g) %>%
#   collect() %>%
#   filter(Base %in% ws_oct,
#          Group == 'CellType (Predict)',
#          Against == 'All') %>%
#   select(Gene, Organism, log2FoldChange, Base)
# 
# consist_diff_wide <- consist_diff_long %>% 
#   mutate(GO = paste0(Base, '_', Organism)) %>% 
#   select(-Organism, -Base) %>% 
#   pivot_wider(values_from=log2FoldChange, names_from = GO)
# consist_diff_celltype <- colnames(consist_diff_wide[,-1])
# consist_diff_wide <- consist_diff_wide %>% data.frame()
# row.names(consist_diff_wide) <- consist_diff_wide$Gene
# 
# consist_diff_wide <- consist_diff_wide[,-1]
# 
# 
# 
# consist_diff_organism <- str_extract(consist_diff_celltype, '_\\w+ \\w+') %>% gsub('_','',.)
# consist_diff_cts  <- str_extract(consist_diff_celltype, '.*_') %>% gsub('_','',.)
# consist_diff_genes <- row.names(consist_diff_wide) %>% gsub(' \\(.*','',.)
# 
# consist_diff_base <- consist_diff_genes %>% as_tibble() %>% rename(Symbol = value) %>% left_join(consist_diff %>% select(Symbol, Base) %>% unique()) %>% pull(Base)
# pmid0 and pmid4 is just GENE
# (t.test(pmid0 %>% map(function(x) sum(!is.na(x))) %>% unlist(), pmid4 %>% map(function(x) sum(!is.na(x))) %>% unlist()))$p.value
query0 <- cbind(consist_diff_genes, consist_diff_base) %>% as_tibble() %>%
  left_join(queries, by = c('consist_diff_genes' = 'Symbol', 'consist_diff_base' = 'Base')) %>%
  mutate(citation = case_when(
    query0 > mean(pmid4 %>% map(function(x) sum(!is.na(x))) %>% unlist()) ~ "Many",
    query0 > median(pmid4 %>% map(function(x) sum(!is.na(x))) %>% unlist()) ~ "Some",
    TRUE ~ "Few")) %>%
  pull(citation)

# pmid2 and pmid3 is "AND RETINA"
# (t.test(pmid2 %>% map(function(x) sum(!is.na(x))) %>% unlist(), pmid3 %>% map(function(x) sum(!is.na(x))) %>% unlist()))$p.value
query2 <- cbind(consist_diff_genes, consist_diff_base) %>%
  as_tibble() %>%
  left_join(queries, by = c('consist_diff_genes' = 'Symbol', 'consist_diff_base' = 'Base')) %>%
  mutate(citation = case_when(
    query2 > mean(pmid3 %>% map(function(x) sum(!is.na(x))) %>% unlist()) ~ "Many",
    query2 > median(pmid3 %>% map(function(x) sum(!is.na(x))) %>% unlist()) ~ "Some",
    TRUE ~ "Few")) %>%
  pull(citation)
#rand_col <-  circlize::colorRamp2(c('0', '1', '2'), c("white", "gray", "black"))
citation_col <-c('white','gray','black')
names(citation_col) <-  c('Few','Some','Many')




consist_diff_hm <- ComplexHeatmap::Heatmap(consist_diff_wide,
                                           column_labels = consist_diff_organism,
                                           column_split = consist_diff_cts,
                                           row_split = consist_diff_base,
                                           show_row_dend = FALSE,
                                           column_title_rot = 90,
                                           name = 'log2FC', use_raster = TRUE,
                                           row_title_rot = 0) +
  ComplexHeatmap::Heatmap(as.character(query0),
                          row_split = consist_diff_base,
                          col = citation_col,
                          row_labels = consist_diff_genes,
                          show_heatmap_legend = FALSE,
                          column_labels = "Pubmed Gene Citations",
                          column_names_side = 'top') +
  ComplexHeatmap::Heatmap(as.character(query2),
                          row_split = consist_diff_base,
                          col = citation_col,
                          row_labels = consist_diff_genes,
                          name = 'Citation Count',
                          column_labels = "Pubmed Gene + \"Retina\" Citations",
                          column_names_side = 'top')
ComplexHeatmap::draw(consist_diff_hm, padding = unit(c(0, 0, 1.5, 0), "in"), row_title = "Marker Genes")

#toc()
```

## Tease apart the similar neurogenic and RPC cell types

```{r neuro, fig.width=30, fig.height=8, echo = F, message = F, fig.cap=neuro_cap}
neuro_cap <- fig_cap(name='neuro', caption = 'A. Genes proposed by Lu et al. as being significantly differentially expressed during the RPC to neurogenic differentiation in human. We plot the differential expression (log2(fold change) directly between RPC and neurogenic cells. The standard error (calculated by DESeq2) is plotted. Genes with a padj < 0.1 and padj < 0.01 are labelled with "*" and "**", respectively. B. Genes from Lu et al. that we replicated with the scEiaD data. C. Novel genes we propose as being differentially expressed between RPC and neurogenic cells')
# tic()
# pairwise_neurogenic_test<-  scEiaD_2020_v01 %>% 
#   tbl("diff_testing") %>% 
#   filter(Against  == "RPC",   
#          Group == 'CellType (Predict)', 
#          Base == "Neurogenic Cell") %>% 
#   group_by(Base, Against, Gene) %>% 
#   summarise(`mean log2FC` = mean(log2FoldChange), `mean padj` = mean(padj), `sig count` = sum(pvalue < 1e-2), `na count` = sum(is.na(log2FoldChange))) %>% 
#   arrange(-`mean log2FC`) %>%   
#   collect() %>% 
#   filter(!is.na(`mean log2FC`))
# 
# 
# # mutate(GENEID = str_extract(Gene, 'ENS\\w+')) %>% 
# # left_join(gene_tab, by = 'GENEID') %>% 
# # relocate(SYMBOL, GENEID) %>% 
# # dplyr::rename(Symbol = SYMBOL, ID = GENEID) %>% 
# # arrange(-`mean log2FC`)
# 

lu_et_al <- c("PTTG1", "SCG3", "SHD", "SOX11", "ONECUT2", "VSX1", "ASCL1", "TCF4", "SOX4", "RORB", "GADD45G", "MYBL1", "MFAP4", "HES6", "NEUROD1", "C8orf46", "OTX2", "OLIG2", "INSM1", "PCBP4", "NEUROD4", "FAM131C", "HES4", "HES5", "HMGA1", "YBX1", "GADD45A", "ATOH7", "DLX2", "DLX1", "SFRP1", "MYB", "NFIB", "SOX9", "ZFP36L1", "FOS", "JUN", "ZFP36L2", "SOX2", "HES1", "EGR1", "JUNB", "NR4A1", "ETV5", "FOXP1", "ID1", "ID2", "ID3", "ID4", "FOXP2")

g_lu <- scEiaD_2020_v01 %>% tbl('genes') %>% collect() %>% filter(grepl(paste(paste0('^',lu_et_al), collapse = ' |'), Gene)) %>% pull(Gene)



down <- pairwise_neurogenic_test %>% filter(`sig count` > 1) %>% arrange(`mean padj`) %>% filter(`mean log2FC` < 0, `mean padj` < 0.05) %>% mutate(d = `mean log2FC` - lfcSE) %>% arrange(-d) %>% pull(Gene)
up <-  pairwise_neurogenic_test %>% filter(`sig count` > 1) %>% arrange(`mean padj`) %>% filter(`mean log2FC` > 2, `mean padj` < 0.05) %>% mutate(d = `mean log2FC` - lfcSE) %>% arrange(-d) %>% pull(Gene)

source('src/make_exp_plot.R')

input <- list()
input$exp_plot_groups <- c('study_accession')
input$exp_plot_facet <- c('CellType_predict', 'organism')
input$exp_plot_genes <- c(down, up[!up %in% g_lu])
input$exp_filter_cat <- 'CellType_predict'
#input$exp_filter_on <- c('AC/HC Precursor', 'Amacrine Cell', 'Astrocyte', 'Bipolar Cell', 'Cone', 'Early RPC',' Horizontal Cell','Late RPC','Microglia','Muller Glia', 'Neurogenic Cell', 'Photoreceptor Precursor','Retinal Ganglion Cell','Rod','Rod Bipolar Cell','RPC','RPE' )
input$exp_filter_on <- c("Neurogenic Cell", "RPC" )
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
#input$exp_plot_ylab <- '% of Cells Detected'
#input$exp_plot_groups <- c('organism')
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 25
input$flip_facet <- FALSE
input$figure <- TRUE

exp_plot_new <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) 

input$exp_plot_genes <- up[up %in% g_lu]
exp_plot_existing <- make_exp_plot(input, scEiaD_2020_v01, meta_filter)
a_legend <- get_legend(exp_plot_new)
neuro_plot <- plot_grid(plot_grid(exp_plot_existing + theme(legend.position="none") + xlab(''),
                                  exp_plot_new + theme(legend.position="none") + ylab('') + xlab(''), 
                                  labels = c('b','c'),
                                  rel_widths = c(0.6,1)),
                        a_legend, ncol = 1, rel_heights = c(1,0.2))


lu_long <-  scEiaD_2020_v01 %>% 
  tbl("diff_testing") %>% 
  filter(Gene %in% g_lu,
         Against  == "RPC",   
         Group == 'CellType (Predict)', 
         Base == "Neurogenic Cell") %>% 
  #group_by(Base, Against, Gene) %>% 
  #summarise(`mean log2FC` = mean(log2FoldChange), `mean padj` = mean(padj), `sig count` = sum(pvalue < 1e-2), `na count` = sum(is.na(log2FoldChange))) %>% 
  #arrange(-`mean log2FC`) %>%   
  collect() %>% 
  mutate(Gene = str_extract(Gene, '\\w+ ') %>% 
           gsub(' ', '', .)) %>% 
  mutate(Gene = factor(Gene, levels = rev(lu_et_al)))


#order <- lu_et_al %>% enframe() %>% mutate(Gene = value) %>% left_join(g_lu %>% enframe() %>% mutate(full = value, Gene = str_extract(value, '\\w+ ') %>% gsub(' ', '', .)) %>% select(-value), by = 'Gene') %>% data.frame() %>% filter(!is.na(full)) %>% pull(full)


lu_rep_plot <- lu_long %>% 
  mutate(Significance = case_when(padj < 0.01 ~ '**', padj < 0.1 ~ '*', TRUE ~ '')) %>% 
  ggplot(aes(x=Gene,y=log2FoldChange,  label = Significance)) + 
  geom_bar(stat='identity', position = position_dodge2(), fill = 'gray') + 
  geom_errorbar(aes(ymin = log2FoldChange - lfcSE,
                    ymax = log2FoldChange + lfcSE), width = 0.5,
                position = position_dodge2()) +
  ylab('Neurogenic / RPC Gene Expression (log2 Fold Change)') +
  coord_flip() + 
  cowplot::theme_cowplot()  + geom_text(aes(y=7), position = position_dodge2(width = 1)) + facet_wrap(~Organism)
#g_lu[!g_lu %in% lu_long$Gene]


cowplot::plot_grid(lu_rep_plot, neuro_plot, rel_widths = c(0.6,1), labels = c('a'))
```

While the well supported markers we propose clearly define most of the cell types, we see that the neurogenic cells and RPCs have very similar expression patterns across the gene sets. These cell types are closely related as the neurogenic cells are derived from the multi-potent RPC to later specify neural cell types. We use the set of genes identified by Lu et al. [@luSingleCellAnalysisHuman2020b] in their study of human retinal differentiation to determine whether our scEiaD database retains the same trends (`r fig_cap(name = 'neuro', display = 'cite')`A). As expected, we see several patterns. First, we see most of the gene expression changes between RPC and neurogenic are matched between human. We further show that these genes also show the same trends in mouse datasets. Next we see that the direction of the changes largely matches Lu et al (`r supFig_cap(name='lu_et_al', display = 'cite')`). Finally, we see a small number of genes are confirmed in our database as consistently differentially expressed with a padj < 0.1. 

To determine whether we can propose any new genes that distinguish RPC and neurogenic states we use our DESeq2 differential expression contrast that directly compares these two cell types. We find `r pairwise_neurogenic_test_ungrouped %>% filter(abs(log2FoldChange) > 2, padj < 0.05, Organism == 'Homo sapiens') %>% nrow()` human genes and `r pairwise_neurogenic_test_ungrouped %>% filter(abs(log2FoldChange) > 2, padj < 0.05, Organism == 'Mus musculus') %>% nrow()` mouse genes that have an absolute log2 fold change greater than two and a adjusted p-value below 0.05 (`r supTab_cap(name = 'neuro_table', display = 'cite')`). To identify candidate genes that are consistently differentially expressed at this transitional state in mouse and human we apply similar logic that we used in identifying well supported marker genes across the retina cell types. We filter to genes with a padj less than 0.05 in both human and mouse, a mean log2 fold change greater than two or less than negative one (`r fig_cap(name = 'neuro', display = 'cite')`B and C). We find one gene, ZBTB38, that meets this criteria and drops in expression when comparing RPC to neurogenic. We find `r up %>% length()` genes that increase in expression from RPC to neurogenic, `r up[!up %in% g_lu] %>% length()` of which are not previously identified by Lu et al (`r fig_cap(name = 'neuro', display = 'cite')`C). We replicate `r up[up %in% g_lu] %>% str_extract(., '\\w+') %>% paste(., collapse = ', ') %>% str_to_title()` from Lu et al (`r fig_cap(name = 'neuro', display = 'cite')`B) as differentially expressed between neurogenic and RPC. Futhermore we show in `r supFig_cap(name = 'PLAE_screenshots_supplemental', display = 'cite')`a how this figure can be replicated directly in our web app. 

## Our pan-study macula vs peripheral cone test replicates a published macaque finding and proposes six new markers

```{r region_fig, fig.width=18, fig.height=3,  echo = F, message = F, fig.cap = region_fig_cap}
region_fig_cap <- fig_cap(name='regino_fig', caption = 'Plot of markers distinguishing human macula or peripheral punch derived cones.')

source('src/make_exp_plot.R')
#load('data/pseudoBulk_cone_region_files.Rdata')
region_table <- read_csv('supplemental_files/pseudoBulk_cone_region_table.csv.gz')
input <- list()
input$exp_plot_groups <- c('study_accession')
input$exp_plot_facet <- c('CellType_predict', 'organism', 'retina_region')
input$exp_plot_genes <- c('PRPH2 (ENSG00000112619)')
input$exp_filter_cat <- c('CellType_predict', 'organism')
#input$exp_filter_on <- c('AC/HC Precursor', 'Amacrine Cell', 'Astrocyte', 'Bipolar Cell', 'Cone', 'Early RPC',' Horizontal Cell','Late RPC','Microglia','Muller Glia', 'Neurogenic Cell', 'Photoreceptor Precursor','Retinal Ganglion Cell','Rod','Rod Bipolar Cell','RPC','RPE' )
input$exp_filter_on <- c("Cone", "Homo sapiens" )
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
#input$exp_plot_ylab <- '% of Cells Detected'
#input$exp_plot_groups <- c('organism')
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 10
input$flip_facet <- FALSE
input$figure <- FALSE
input$figure_fovea <- TRUE

prph2 <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') +  theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <-  'RS1 (ENSG00000102104)';rs1 <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') +  theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 

input$exp_plot_genes <-  'GNGT1 (ENSG00000127928)';gngt1 <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') +  theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
#input$exp_plot_genes <-  'RS1 (ENSG00000102104)';rs1 <- make_exp_plot(input, scEiaD_2020_v01, meta_filter)


diff_region_cone_genes <- region_table %>% filter(padj < 0.01) %>% 
  arrange(log2FoldChange) %>% 
  mutate(id = paste0(Gene, ' (', Ensembl, ')')) %>% 
  pull(id)
input$exp_filter_on <- c("Cone", "Homo sapiens" )
input$exp_plot_genes <- diff_region_cone_genes[1]; a <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') +  theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[2]; b <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[3]; c <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[4]; d <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + ylab('') + xlab('') + theme(axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[5]; e <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[6]; f <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[7]; g <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 

region_legend <- get_legend(d)

plot_grid(plot_grid(a,b,c,d + theme(legend.position="none"),e,f,g, nrow = 1),
          region_legend, 
          rel_heights = c(1,0.2), 
          nrow = 2)
```


While we have provided data in many forms via the web app at plae.nei.nih.gov, there are many more dissections of the data that can be performed. To facilitate custom testing with the scEiaD database we provide, as an example, a human-specific macula versus peripheral cone differential testing analysis document as supplementary file "pseudobulk_cone_region.freeze01.html". In our analysis document we use the Seurat object containing the full scEiaD dataset we provide at plae.nei.nih.gov to build a custom pseudobulk counts matrix containing only cones from human studies that were taken in a region specific manner. This left us with 26678 cells across six studies. Yan et al. previously proposed Prph2 and Rs1 as markers to distinguish macula and peripheral located human cones[@yanCellAtlasHuman2020a] and Peng et al. proposed Calb1 and Gngt1 as cone region markers in macaque [@pengMolecularClassificationComparative2019a]. 

Our differential testing found that while Prph2, Rs1, and Gngt1 had expression differences matching that the findings Yan et al. and Peng et al. published, they failed to replicate in our test with pvalue and padj being above 0.1 for all three. In contrast Calb1 did replicate strongly with the sixth lowest padj (0.00721) across our entire test and a log2 fold change of 2.87. We find six more genes with a padj < 0.01: Pde6g, Akap12, Hspb1, Gadd45b, Cc2d2a, and RP1l1 (`r fig_cap(name = 'region_fig', display = 'cite')` and `r supTab_cap(name = 'region_table', display = 'cite')`). None of these candidate cone region marker genes display regionally different expression in rods (`r supFig_cap(name = 'rod_region_fig', display = 'cite')`).

## Limitations of scEiaD 

The longer term goal is to continue adding datasets to the scEiaD at plae.nei.nih.gov until we have three or more independent datasets per species and most ocular cell types. Earlier we proposed well supported cell types. Right now, most of these well supported cell types are in the retina. The RPE behind the retina, the cornea and lens in front, and the outflow tract and supporting musculature still could use more independent studies. Many common human diseases underlie the RPE (e.g. AMD) and the cornea (e.g. Keratoconus) and these tissues will remain a focus for the next update of our database. While we have curated a large number of cell types across the eye, many of these cell types have "sub" cell types. For examplethe human has three types of cones that are optimized for short, medium, and long wavelengths. More dramatically, Yan et al. have proposed over sixty types of mouse amacrine cell types [@yanMouseRetinalCell2020a]. While we have found it fairly straightforward to use machine learning to transfer the broad (e.g. rod, cone, amacrine) cell type labels across all datasets, we have found it very challenging to transfer the "sub" cell type labels. This is largely because while we have high diversity of studies for the broad cell types, this is not true for most of the sub celltypes. We believe more datasets that carefully dissect the sub-celltypes may be necessary to confidently propose the sub-celltypes of the retina. Finally, while our scVI based model can theoretically be used by an outside group to transfer their internal data into the scEiaD latent space, it is very challenging to implement as the the software landscape is still rapidly changing and it is very difficult to architect consistent outputs over time with regularly changing software versions. We are investigating simpler methods to enable transfer of knowledge from scEiaD to outside data. 


# Conclusions


## High feature web app provides quick access to over one million transcriptomes across `r meta_filter$study_accession %>% unique %>% length` datasets

We have improved on the scEiaD dataset we published in Swamy et al. [@swamyBuildingMegaSinglecell2021] by adding another species (chicken), new ocular tissues, a human non-ocular reference set; in total `r meta_filter$Citation %>% unique() %>% length() - meta_filter_v0$Citation %>% unique() %>% length()` new single cell datasets. We re-structured the data into a custom database to enable our custom built high performance web app to quickly access the data. This democratizes access to over a million cells across `r meta_filter$study_accession %>% unique() %>% length()` studies to anyone with a web browser. We recognize that while we have built a powerful visualization platform we cannot anticipate all uses and for those researchers who wish to make custom queries of the data, we provide the data underlying plae in a wide variety of formats, including anndata and Seurat objects for each study, count tables, cell level metadata, and the DESeq2 based differential testing across three species. We provide a short example of how a custom query could work by running a custom differential test between macula and peripheral derived human cones and demonstrate how this can both validate existing proposed genes and suggest new ones. 

## Diversity of studies enables reproduction and extension of cell type gene expression findings

A great many groups have proposed cell atlases. These range from efforts focusing on a subset of cells from a single layer of a tissue (for example [@tranSingleCellProfilesRetinal2019a] ) to a huge consortium encompassing multiple organs [@consortiumTabulaSapiensSingle2021]. We propose a complementary approach wherein we use the research data from the community at large to create a meta-atlas. This approach enables a small group to produce an atlas which rivals the size from massive multi-group consortiums. Furthermore the inherent variability between the data from independent groups can be used as a feature to get insight into whether a gene cell type signal is stable across an entire community's datasets. We briefly demonstrate how the high study diversity in scEiaD can be used to validate proposed RPC to neurogenic gene expression changes. 

# Supplementary Figures and Tables


```{r decontX, fig.height = 15, fig.width = 20, echo = F, message = F, fig.cap=decontX_cap}
decontX_cap <- supFig_cap(name='decontX', caption = 'DecontX *in silico* ambient RNA contamination tool substantially removes Rhodopsin expression in non-rod cells while retaining high expression rods. First shown (a) is Rhodopsin expression across the scEiaD v0 dataset, without DecontX optimization. Note Rhodopsin expression is noticeable in many cell types beyond the rods. Next (b) is the scEiaD v1 dataset with DecontX optimization. Rhodopsin is nearly exclusively expressed in labelled rod cells. Despite the computational ambient RNA removal, Rhodopsin expression remains high in the Rods.')
source('src/make_exp_plot.R')

input <- list()
input$exp_plot_groups <- c('organism')
input$exp_plot_facet <- c('CellType_predict')
input$exp_plot_genes <- c('RHO (ENSG00000163914)')
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 50
input$flip_facet <- FALSE
input$figure <- FALSE
input$figure_fovea <- FALSE


v0_rho <- make_exp_plot(input, scEiaD_2020_v00, meta_filter_v0)
v1_rho <-  make_exp_plot(input, scEiaD_2020_v01, meta_filter)
input$exp_filter_min_cell_number <- 1000
v1.1_rho <- make_exp_plot(input, scEiaD_2020_v01, meta_filter %>% filter(CellType_predict %in% unique(meta_filter_v0$CellType_predict)))
plot_grid(v0_rho, v1_rho, ncol = 1, labels = c('a','b'), rel_heights = c(1,1.5))
```

```{r sup_scEiaD_data_overview, fig.height = 16, fig.width = 20, echo = F, message = F, fig.cap=sup_scEiaD_data_overview_cap}

num1 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num2 <- meta_filter %>% filter(Organ != 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num3 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(CellType) %>% unique() %>% length()
sup_scEiaD_data_overview_cap <- supFig_cap(name = '_supscEiaD_data_overview', caption = glue::glue("Distribution of tissues and cell types in scEiaD. A. tabulation of the number of cells present across {num1} ocular and {num2} non-ocular tissues, B. Number of studies for each ocular tissue. C. Number of cells present across {num3} curated ocular-derived cell types. D. Proportion of cell types (predicted) across each study (split by species and tissue)."))

# organ
meta_filter <- meta_filter %>% mutate(organism = factor(organism, levels = meta_filter$organism %>% unique()))
a <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Body') %>% 
  filter(Tissue != 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  # scale_y_continuous(breaks = c(2500, 5000, 7500, 10000, 12500, 84000)) +
  # ggbreak::scale_y_break(c(13000, 82000)) +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

aa <- meta_filter %>% 
  filter(Tissue == 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

# eye studies
b <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = case_when(Tissue == "Choroid" ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  group_by(Organ, Tissue,organism) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) +
  xlab('')

# eye studies samples
c <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  select(Organ, Tissue,organism, study_accession) %>%
  unique() %>%
  group_by(Organ, Tissue, organism) %>% 
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0))  + xlab('')

# retina cell types
d <- meta_filter %>% filter(Organ == 'Eye') %>% 
  mutate(CellType_predict = case_when(is.na(CellType_predict) ~ 'Unlabelled', TRUE ~ CellType_predict)) %>% 
  group_by(CellType_predict, study_accession, organism) %>%
  summarise(Count = n()) %>% filter(Count > 10) %>% ungroup() %>% 
  ggplot(aes(x=CellType_predict,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')


a_aa_b <- plot_grid(a + scale_y_continuous(guide = guide_axis(check.overlap = TRUE)),
                    aa + scale_y_continuous(guide = guide_axis(check.overlap = TRUE)),
                    b + scale_y_continuous(guide = guide_axis(check.overlap = TRUE)), 
                    ncol = 1, align = 'hv', rel_heights = c(10, 1.5,3.5), labels = 'a')
org_legend <- get_legend(d)
c_d <- plot_grid(c,d + theme(legend.position = "none"), rel_heights = c(3,10), ncol = 1, align= 'hv', axis = 'l',labels = c('b','c'))


# make E now

# map color as plae app does
categorical_columns <- c("Phase","batch","study_accession","library_layout","organism","Platform",
                         "Covariate","CellType","CellType_predict","TabulaMurisCellType","TabulaMurisCellType_predict",
                         "GSE","Summary","Design","Citation","PMID","Stage","cluster",
                         "Doublet","TechType", "SubCellType", 'subcluster', 'Age', "retina_region",
                         'Tissue','Organ', 'Source','sample_accession')
#"SubCellType" and subcluster are problems
meta_filter <- meta_filter %>% mutate(Age = as.character(Age), SubCellType = tidyr::replace_na(SubCellType, 'None'),
                                      subcluster = as.character(subcluster))


map_color <- function(column, meta_filter){
  master_colorlist <- c( pals::glasbey(), pals::alphabet(), pals::cols25()[1:23], pals::alphabet2()) %>% unique()
  values <- meta_filter %>% pull(!!column) %>% unique %>% sort
  if(length(values) > length(master_colorlist) ){
    r= round(length(values) / length(master_colorlist)) +1
    master_colorlist <- rep(master_colorlist, r)
  }
  
  colors <- master_colorlist[1:length(values)]
  return(tibble(meta_category = column,value = values, color=colors))
  
}

cat_to_color_df <- lapply(categorical_columns, function(col) map_color(col, meta_filter)) %>% bind_rows()
# now roll through and set each value to the same color across CellType and CellType (predict)
# e.g. RPE will be one color in CellType and CellType (Predict)
celltype_x_color_map <- cat_to_color_df %>% filter(meta_category %in% c("CellType", "CellType_predict")) %>% group_by(value) %>% summarise(c2 = head(color,1))
cat_to_color_df <- cat_to_color_df %>% left_join(celltype_x_color_map) %>% mutate(color = case_when(!is.na(c2) ~ c2, TRUE ~ color)) %>% select(-c2)


o_ct <- meta_filter %>% filter(Tissue == 'Retina') %>% pull(CellType)
p_data <- meta_filter %>% 
  #mutate(CellType_predict = Tissue) %>% 
  #mutate(CellType_predict = case_when(!CellType_predict %in% ws_oct ~ 'Other',
  #                                    TRUE ~ CellType_predict)) %>% 
  filter(Organ == 'Eye') %>% 
  #mutate(Citation = paste(str_trunc(Citation, width = 20), str_trunc(Covariate, width = 12, side = 'left',ellipsis = ''))) %>% 
  mutate(Citation = str_trunc(Citation, width = 20)) %>% 
  group_by(Citation, organism, Tissue, CellType_predict) %>% 
  summarise(Count = n()) %>% 
  mutate(Proportion = Count/sum(Count) * 100) 

meta_column <- 'CellType_predict'
cur_color_df <- cat_to_color_df %>%
  filter( meta_category %in%  meta_column,
          value %in% p_data[[meta_column]]) %>% distinct

color_list <- cur_color_df %>% pull(color)
# replaced join with this for speed
k <- cur_color_df$color
names(k) <- cur_color_df$value
np_color <- {k[p_data[[meta_column]] ]} %>% paste0(., '33')
names(np_color) <- NULL
color_data <- cur_color_df  %>%
  select(value) %>%
  mutate( x=0)
names(color_list) <- color_data$value

proportion_bar_plot <- p_data %>% ggplot(aes(y=Citation,x=Proportion,fill=CellType_predict)) + geom_bar(stat= 'identity') +
  scale_fill_manual(name= meta_column,
                    values = color_list) + 
  ggforce::facet_col(vars(organism, Tissue), scale = 'free', space = 'free') +
  cowplot::theme_cowplot() +
  guides(fill=guide_legend(ncol =1))


e <- plot_grid(proportion_bar_plot+ theme(legend.position = "none"), labels = c('d'))
ct_legend <- get_legend(proportion_bar_plot)
fig_overview_legends <- plot_grid(org_legend, ct_legend,ncol=1)
plot_grid(a_aa_b, c_d , e ,
          fig_overview_legends, ncol =4, rel_widths = c(1,1.6, 2.2, 0.8))
#toc()()
```

```{r undersupported_genes, fig.height=48, fig.width=8, echo = FALSE}
cone_undersupported <- queries %>% filter(query2 <= 1) %>% arrange(-`mean log2FC`) %>% filter(!grepl("Endo|Macro|Micro|Endo", Base)) %>% group_by(Symbol) %>% summarise(Count = n(), Base = paste0(Base, collapse = ', '), l2FC = mean(`mean log2FC`)) %>% filter(Count == 1) %>% filter(Base == 'Rod') %>% arrange(-l2FC)





input <- list()
input$exp_plot_groups <- c('study_accession')
input$exp_plot_facet <- c('CellType_predict')
input$exp_plot_genes <- scEiaD_2020_v01 %>% tbl('Genes') %>% collect() %>% mutate(Symbol = str_extract(Gene, '.* \\(') %>% gsub(' \\(','',.)) %>% filter(Symbol %in% cone_undersupported$Symbol) %>% pull(Gene)
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 50
input$flip_facet <- FALSE
input$figure <- FALSE
input$figure_bloop <- TRUE
input$figure_fovea <- FALSE


make_exp_plot(input, scEiaD_2020_v01, meta_filter)

```
```{r PLAE_screenshots_supplemental, echo = F, message = F, fig.cap = PLAE_screenshots_supplemental_cap}
PLAE_screenshots_supplemental_cap <- supFig_cap(name='PLAE_screenshots_supplemental', caption = 'a. Recreation of Figure 4 directly in the plae.nei.nih.gov web app. b. Identify potential bipolar cell markers with the precomputed differential testing and in c. visualize the top five genes with a Heatmap.')
knitr::include_graphics('fig/plae_sup_fig_screenshots.png')

```


```{r lu_et_al, echo = F, message = F, fig.cap = lu_et_al_cap}
lu_et_al_cap <- supFig_cap(name='lu_et_al', caption = 'Supplemental figure 2K from Lu et al showing their candidate RPC / Neurogenic differentially expressed genes')
knitr::include_graphics('fig/lu_et_al_fig2k_screenshot.png')

```


```{r rod_region_fig, fig.width=20, fig.height=10,  echo = F, message = F, fig.cap = rod_region_fig_cap}
rod_region_fig_cap <- supFig_cap(name = 'rod_region_fig', caption = 'Human cone fovea - peripheral differentially expressed genes are not differentially expressed in human rods.')

# 
# rods
input <- list()
input$exp_plot_groups <- c('study_accession')
input$exp_plot_facet <- c('CellType_predict', 'organism', 'retina_region')
input$exp_plot_genes <- c('PRPH2 (ENSG00000112619)')
input$exp_filter_cat <- c('CellType_predict', 'organism')
#input$exp_filter_on <- c('AC/HC Precursor', 'Amacrine Cell', 'Astrocyte', 'Bipolar Cell', 'Cone', 'Early RPC',' Horizontal Cell','Late RPC','Microglia','Muller Glia', 'Neurogenic Cell', 'Photoreceptor Precursor','Retinal Ganglion Cell','Rod','Rod Bipolar Cell','RPC','RPE' )
input$exp_filter_on <- c("Rod", "Homo sapiens" )
input$exp_plot_ylab <- 'Mean log2(Counts + 1)'
#input$exp_plot_ylab <- '% of Cells Detected'
#input$exp_plot_groups <- c('organism')
input$exp_plot_col_num <- 16
input$exp_filter_min_cell_number <- 10
input$flip_facet <- FALSE
input$figure <- FALSE
input$figure_fovea <- TRUE
input$exp_plot_genes <- diff_region_cone_genes[1]; a <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') +  theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[2]; b <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[3]; c <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[4]; d <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + ylab('') + xlab('') + theme(axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[5]; e <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[6]; f <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 
input$exp_plot_genes <- diff_region_cone_genes[7]; g <- make_exp_plot(input, scEiaD_2020_v01, meta_filter) + xlab('') + ylab('') + theme(legend.position="none",axis.ticks.x=element_blank(), axis.text.x=element_blank() ) 

region_legend <- get_legend(d)

plot_grid(plot_grid(a,b,c,d + theme(legend.position="none"),e,f,g, nrow = 1),
          region_legend, 
          rel_heights = c(1,0.2), 
          nrow = 2)
```



```{r cone_region_rep, fig.width=8, fig.height=10,  echo = F, message = F, fig.cap = cone_region_rep_cap}
cone_region_rep_cap <- supFig_cap(name = 'cone_region_rep', caption = 'Genes identified by Peng et al. and Yan et al. as macula / peripheral differentially expressed in cones.')
cowplot::plot_grid(prph2, rs1,gngt1, ncol = 1)
```



```{r studies_table, fig.height = 15, fig.width = 20, echo = F, message = F, results = 'asis',fig.cap=studies_table_cap}
# 1

studies_table_cap <- supTab_cap(name = 'studies_table', caption = 'Studies, tissues, and number of cells across scEiaD')
cat("<div id=\"tbl:studies_table\">")
meta_filter %>% group_by(Citation, PMID, organism, study_accession) %>% summarise(Count = n(), Tissues = paste(unique(Tissue), collapse = ', ')) %>% flextable() %>% autofit() %>% fit_to_width(7)

#Output Caption (* * for italics)
cat(paste0("\n*",studies_table_cap,"*\n"))
# Output </div>
cat("</div>")
```


```{r diff_tab_sup_tab, echo = F, message = F,  echo = F, message = F, results = 'asis', fig.cap=diff_tab_sup_tab_cap}
# 2
diff_tab_sup_tab_cap <- supTab_cap(name='diff_tab_sup_tab', caption = 'Counts of significantly differentially expressed genes for each cell type across the organisms')
cat("<div id=\"tbl:diff_tab_sup_tab\">")
# tic()
# # number of celltype (predict) - organism - study counts (with >=50 cells)
# # ws == Well Supported
# ws_table <- meta_filter %>%  
#   filter(Organ == 'Eye') %>% 
#   group_by(organism, CellType_predict, study_accession) %>% 
#   count() %>% 
#   filter(n >= 50) %>% 
#   ungroup() %>% 
#   group_by(organism, CellType_predict) %>% 
#   count() %>% 
#   pivot_wider(values_from = n, names_from = organism) %>% 
#   filter(!is.na(CellType_predict)) %>% 
#   replace(is.na(.), values = 0) %>% 
#   arrange(CellType_predict) %>% 
#   rename(`CellType (Predict)` = CellType_predict) %>% 
#   mutate(`Well Supported` = case_when(`Homo sapiens` >= 2 & `Mus musculus` >= 2 ~ 'Yes',
#                                       TRUE ~ 'No'))
# 
# # fix  occasional NA
# # data made from src/ensdb.R
# diff_tab_big <- read_tsv('data/diff_tab_big.tsv.gz')
# down_tab_big <- read_tsv('data/down_tab_big.tsv.gz')
# gene_tab <- bind_rows(read_tsv('data/hs.ensdb.v86.symbol.geneid.tsv.gz'), read_tsv('data/mm.ensdb.v79.symbol.geneid.tsv.gz'))
# diff_tab <- diff_tab_big %>% 
#   filter(Against == 'All', 
#          Base %in% (meta_filter %>% filter(Organ == 'Eye') %>% 
#                       pull(CellType) %>% unique()))
# 
# 
# diff_tab_sup <- diff_tab %>% 
#   group_by(Base, Organism) %>% 
#   summarise(Count = n()) %>% 
#   rename(`CellType (Predict)` = Base, `Differentially Expressed Gene Count` = Count) %>% 
#   pivot_wider(values_from = `Differentially Expressed Gene Count`, names_from = Organism) %>% replace(is.na(.),0) %>% 
#   as_tibble()
# 
# diff_tab_sup <- bind_rows(
#   diff_tab_sup,
#   ws_table$`CellType (Predict)`[!ws_table$`CellType (Predict)` %in% 
#                                   diff_tab_sup$`CellType (Predict)`] %>% 
#     enframe() %>% 
#     select(`CellType (Predict)` = value) %>%
#     mutate(`Homo sapiens` = 0, `Mus musculus` = 0, `Macaca fascicularis` = 0) %>% 
#     arrange(`CellType (Predict)`)
# )
# 
# write_csv(diff_tab, file = 'supplemental_files/differential_expression_table.csv.gz')

# diff_tab_sup %>% 
#   flextable()
# 
# ws_table %>% 
#   flextable()

diff_tab_barplot <- diff_tab_sup %>% 
  pivot_longer(-`CellType (Predict)`, names_to = 'Organism', values_to = 'Diff Gene Count') %>% 
  ggplot(aes(x=`CellType (Predict)`, y = `Diff Gene Count`, fill = Organism )) + 
  geom_bar(stat='identity', position = 'dodge') + 
  cowplot::theme_cowplot() + coord_flip()

#toc()


diff_tab_sup %>% flextable() %>% autofit() %>% fit_to_width(7)

#Output Caption (* * for italics)
cat(paste0("\n*",diff_tab_sup_tab_cap,"*\n"))
# Output </div>
cat("</div>")
```


```{r well_supported_celltypes_tab, echo = F, message = F, results = 'asis', fig.cap=well_supported_celltypes_tab_cap}
# 2
well_supported_celltypes_tab_cap <- supTab_cap(name='well_supported_celltypes_tab', caption = 'Counts of independent studies for each cell type across the organisms')
cat("<div id=\"tbl:well_supported_celltypes_tab\">")
ws_table %>% flextable() %>% autofit() %>% fit_to_width(7)

#Output Caption (* * for italics)
cat(paste0("\n*",well_supported_celltypes_tab_cap,"*\n"))
# Output </div>
cat("</div>")
```


```{r neuro_table, echo = F, message = F,  echo = F, message = F, results = 'asis',fig.cap=neuro_table_cap}
# 4
neuro_table_cap <- supTab_cap(name='neuro_table', caption = 'Differentially expressed genes between RPC and neurogenic cells in human and mouse. Genes filtered with a abs(log2FoldChange) > 2 and a padj < 0.01.')
cat("<div id=\"tbl:neuro_table\">")
pairwise_neurogenic_test_ungrouped %>% filter(padj < 0.05) %>% filter(log2FoldChange > 2 | log2FoldChange < 0) %>% flextable() %>% autofit() %>% fit_to_width(7)

#Output Caption (* * for italics)
cat(paste0("\n*",neuro_table_cap,"*\n"))
# Output </div>
cat("</div>")
```




```{r region_table,  echo = F, message = F, results = 'asis',fig.cap = region_table_cap}
# 5
region_table_cap <- supTab_cap(name = 'region_table', caption = 'Macula versus peripheral cone differential gene testing table')
cat("<div id=\"tbl:region_table\">")
region_table %>% filter(padj < 0.1) %>% 
  arrange(padj) %>% flextable() %>% autofit() %>% fit_to_width(7)

#Output Caption (* * for italics)
cat(paste0("\n*",region_table_cap,"*\n"))
# Output </div>
cat("</div>")
```

















# Works Cited












