---
title: "R Notebook"
output: html_notebook
---

```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
#knitr::opts_chunk$set(fig.pos = 'p') # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)
library(tidyverse)
library(citr)
library(cowplot)
library(ggrepel)
library(colorspace)
library(flextable)
library(captioner)
library(glue)
data_dir_vGiga <- '/Volumes/McGaughey_S/data/scEiaD//'
data_dir_vPLAE <- '/Volumes/McGaughey_S/data/scEiaD_2022_02/'
#data_dir_vGiga <- '~/data/scEiaD_gigascience/'
#data_dir_vPLAE <- '~/data/scEiaD_2022_02/'
# setup caption-ing
fig_cap <- captioner("Figure")
supFig_cap <- captioner("Supplemental Figure")
tab_cap <- captioner("Table")
supTab_cap <- captioner("Supplemental Table")

library(pool)
library(RSQLite)
meta_filter_v0 <- fst::read_fst(glue("{data_dir_vGiga}/paper_data/2021_03_17_meta_filter.fst")) %>% 
  
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precurs' ~ 'AC/HC Precursor',
                             TRUE ~ CellType))

meta_filter <- fst::read_fst(glue("{data_dir_vPLAE}/meta_filter.fst")) %>% filter(study_accession != 'Bharti_Nguyen_iRPE_2D_3D') %>% 
  
  mutate(CellType_predict = case_when(CellType_predict == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                                      TRUE ~ CellType_predict),
         CellType= case_when(CellType == 'AC/HC_Precursor' ~ 'AC/HC Precursor',
                             TRUE ~ CellType),  
         Tissue = case_when(grepl('Iris', Tissue) ~ 'Iris',
                            TRUE ~ Tissue),
         Tissue = case_when(grepl('^RPE|^Choroid', Tissue) ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Citation = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ 'Swamy VS, Fufa TD, Hufnagel RB, McGaughey DM Building the Mega Single Cell Transcriptome Ocular Meta-Atlas Gigascience 2021 Oct 13;10(10)',
                              TRUE ~ Citation),
         PMID = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ '34651173',
                          TRUE ~ PMID),
         study_accession = case_when(study_accession == 'OGVFB_Hufnagel_iPSC_RPE' ~ 'SRP329495',
                                     TRUE ~ study_accession))

scEiaD_2020_v00 <- pool::dbPool(RSQLite::SQLite(), dbname =  glue('{data_dir_vGiga}/MOARTABLES__anthology_limmaFALSE___5000-transform-counts-universe-batch-scVIprojectionSO-8-0.1-50-5.sqlite'))
scEiaD_2020_v01 <- pool::dbPool(RSQLite::SQLite(), dbname =  glue('{data_dir_vPLAE}/MOARTABLES__anthology_limmaFALSE___4000-counts-universe-study_accession-scANVIprojection-15-5-0.1-50-20__pointRelease01.sqlite'))

```



# Introduction

## single cell RNA-seq allows for a higher resolution view of ocular transcriptomics

Single cell RNA-sequencing dates back to . The first 

## Many published ocular single cell RNA-seq atlases


## Web atlases containing ocular data have substantial limitations

Data difficult to access and cross comparisons impossible

But important as seeing trends trends across different groups strongly demonstrates that 
the signal is biological, not technical. 

## What we did

We published the ocular focused scEiaD v2021 resource in 2021, which was one of the first single cell meta-atlases. To broaden the usage of this powerful database, we restructured the data in a custom sqlite-based database and built a reactive web application around it. As many new studies were published since our last data build, we re-ran the scEiaD pipeline in March of 2022. In aggregate we have collated `r meta_filter$study_accession %>% unique() %>% length()` studies across `r meta_filter$PMID %>% unique() %>% length()` publications, four species, and `r meta_filter$Tissue %>% unique() %>% length()` tissues. We hand-curated `r meta_filter$CellType %>% unique() %>% length()` cell types, totaling `r meta_filter %>% filter(!is.na(CellType)) %>% nrow()` cells and used those ground-truth labels as the basis for a machine learning-based algorithm which labels the remaining cells. Extensive quality control metrics were tuned and applied to remove lower quality cells. In the end, the single cell Eye in a Disk (scEiaD) database contains `r meta_filter %>% nrow()` cells and 1,509,179,347 non-zero gene/cell expression values. 

# Methods

## scEiaD pipeline upgraded and the database re-built

The scEiaD pipeline is structurally the same as in Swamy et al. [@swamy_building_2021]. Very briefly, we collect the raw fastq files and quantify with the kallisto / bustools system [@melsted_barcode_2019; @melsted_modular_2019]. The quantified counts are merged into one matrix that is used for downstream processing with scvitools to batch correct the data and Seurat for quality control, clustering, and 2D UMAP visualization. We use our previous xgboost-based machine learning approach to transfer cell type labels to unlabelled cells [@swamy_building_2021].  

The following alterations were made: First, the cutoff to retain a cell was raised from 200 unique genes quantified to 300 Second, we use the DecontX algorithm [@yang_decontamination_2020] from the celda R package (version 1.9.2) to automatically remove ambient RNA contamination. Third, we altered the procedure for automatically aligning gene names across species (see below for details). Fourth, we use the scANVI (scvitools based) batch correction method, which leverages the known cell type labels in the batch correction process. 

Two custom cutoffs for the minimum required number of unique genes identified per cell were used for the E12 chicken data and the cornea dataset (SRP362101) as an abnormally high number of cells were returned from our default value of 250 (????). We set the cutoffs for these studies to be 1000 and 800, respectively, as these values resulted in approximately the same number of cells being returned as the authors reported in their papers. 

## Cross species gene alignment

As we have four species in scEiaD (*Gallus gallus*, *Macaca fascicularis*, *Mus musculus*, and *Homo sapiens*) that we unify, we must identify shared genes. We found homologs/orthologs by pulling from BioMart database "Ensembl Genes 105" and using *Homo sapiens* as the reference. We then selected orthologous gene names from *Gallus gallus*, *Macaca fascicularis*, and *Mus Musculus*. This generates a table linked by Ensembl gene IDs and gene names. We then use a full join on non-duplicated Ensembl IDs between mouse and human. As we noticed a small number of gene names failed to be linked in this manner we ran another join on the remaining non-aligned genes, using gene name. This procedure gave us 17,769 shared human and mouse genes. To align the other species we again joined on Ensembl gene ID, removing genes where one chicken or macaque gene aligned to multiple human/mouse genes. In cases where multiple chicken or macaque genes aligned to one human/mouse gene, then we aggregated the counts by sum. In cases where there was no corresponding chicken or macaque gene, we filled in zeros. After these steps we had, as expected, 17,769 genes. 

## Metadata curation

Every study brought into scEiaD was hand curated to identify unique biological samples, published paper ID (where available), organ, tissue (e.g. Cornea), source (iPSC based or tissue), single cell platform (e.g. 10Xv2, DropSeq, etc.). We also, where possible, annotated retina region (e.g. fovea), sex, and age or developmental stage. To import individual cell type labels (e.g. Rod, Muller Glia), we wrote custom code for each study that made this table available to link their cell type assignments back to the matching barcoded cell. We curated two types of cell type labels: "CellType" and "SubCellType" where the former is our normalized cell types taken from the published labels. We normalized names for CellType, for example changing "MG" to "Muller Glia" and also dropping more detailed cell type assignments like off and on bipolar cells for Bipolar Cells (as few studies went into this level). We did retain the original published (except for the broader name normalization) labels under SubCellType. 

## Differential gene testing

The scater::aggregateAcrossCells function was used to aggregate gene counts across species, study accession, and [category] where category was either Cluster, CellType, or CellType (Predict) assignments for a cell. If there were fewer than 50 cells in organism - study accession - category combination, they were discarded for the differential test. Any genes with a sum of 0 counts after aggregation were removed. The aggregated matrix was imported into DESeq2 using the DESeqDataSetFromMatrix function with the design given as "~study_accession + [category]". Contrasts were extracted either with the target category (e.g. "Cone") against all remaining or in a pair wise manner ("Cone" vs "Rod") with the DESeq2 "results" function. Genes were considered to be differentially expressed if they have a padj less than $1 x 10^-5 and a abs(log2FoldChange) > 2. To enhance the specificity of the 


## scEiaD structure and web app optimization

The web app at plae.nei.nih.gov can display gene expression in only a few seconds across over a million cells. This speed is only possible due to custom data structures to optimize for data query efficiency. Like eyeIntegration's EiaD database, scEiaD is a sqlite database with tables for the core categories. This allows the app to initialize in about 30 seconds on a cloud server with memory usage under 8 GB. For the gene by cell expression matrix, the data was transformed into a "long" format where there are three columns: Gene, Cell Barcode, and Gene Count Value. This allows for cell - gene level queries to be completed, again with minimal memory usage, in under 0.5 seconds. To generate the aggregated information (in the Dot Plot, Expression Plot, and tables) we pre-calculated all queries by running a dplyr "group by" operation on all column fields. On user given query, the data is further aggregated to level request. This allows for complicated queries to complete in seconds rather than minutes. 

## Data reproducibility

The codebase for scEiaD is available at github.com/davemcg/scEiaD; the commit corresponding to this manuscript is XXXXXXX. There are three relevant Snakemake pipelines which were used to create scEiaD: SnakeQUANT, which was used to quantify the gene / cell expression from the raw fastq files, SnakePOP, which runs the multi-parameter integration methods, and SnakeSCEIAD, which uses the optimal parameters identified in SnakePOP to run the cell type machine learning and differential gene testing that is incorporated into the scEiaD sqlite database for the app. 

# Results

## New studies and improvements to the scEiaD database

Our first version of the single cell Eye is a Disk (scEiaD) database contained `r meta_filter_v0$study_accession %>% unique() %>% length()` studies, three species, `r nrow(meta_filter_v0)` cells, and `r meta_filter_v0$CellType_predict %>% unique() %>% length()` curated cell types [@swamy_building_2021]. We updated the scEiaD resource for the PLAE v0.91 web app in six important ways. First, we added a large chicken retina (*Gallus gallus*) studies. Second we added an ocular outflow tract dataset and two more cornea datasets to enhance coverage across the eye. Third, we added a human pan-body reference scRNA dataset to allow for non-ocular comparisons. Fourth, we used an *in silico* method to remove background gene contamination as we noticed persistent Rhodopsin expression in many non-photoreceptors cells (`r supFig_cap(name = 'decontX', display = 'cite')`). Fifth, we increased the cell retention cutoff for detected unique transcripts per cell from 200 to 300 (CHECK THIS), to improve the database quality. Sixth, we carried over and harmonized several common cell type labels from the Tabula Muris project to more easily provide non eye cell type comparisons.

After the dataset and quality control updates the scEiaD v2022-03-21 dataset now contains `r meta_filter$study_accession %>% unique() %>% length()` studies, four species, `r nrow(meta_filter)` cells, and `r meta_filter$CellType_predict %>% unique() %>% length()` curated cell types (`r fig_cap(name = 'scEiaD_data_overview', display = 'cite')`). This resource is only possible due to the publication of `r meta_filter$PMID %>% unique %>% length()` single cell resources that scEiaD draws from (`r supTab_cap(name = 'studies_table', display = 'cite')`). To recognize these papers, we make the first data table on the loading page of plae.nei.nih.gov list each of these studies along with a direct link to their citation, where possible. 

```{r scEiaD_data_overview, fig.height = 7, fig.width = 5, echo = F, message = F, fig.cap=scEiaD_data_overview_cap}

num1 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num2 <- meta_filter %>% filter(Organ != 'Eye') %>% pull(Tissue) %>% unique() %>% length()
num3 <- meta_filter %>% filter(Organ == 'Eye') %>% pull(CellType) %>% unique() %>% length()
scEiaD_data_overview_cap <- fig_cap(name = 'scEiaD_data_overview', caption = glue::glue("Distribution of tissues and cell types in scEiaD. A. tabulation of the number of cells present across {num1} ocular and {num2} non-ocular tissues, B. Number of studies for each ocular tissue. C. Number of cells present across {num3} curated ocular-derived cell types."))

# organ
meta_filter <- meta_filter %>% mutate(organism = factor(organism, levels = meta_filter$organism %>% unique()))
a <- meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Body') %>% 
  filter(Tissue != 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  # scale_y_continuous(breaks = c(2500, 5000, 7500, 10000, 12500, 84000)) +
  # ggbreak::scale_y_break(c(13000, 82000)) +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

aa <- meta_filter %>% 
  filter(Tissue == 'Brain Choroid Plexus') %>% 
  group_by(Organ, Tissue,organism, study_accession) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  mutate(organism = factor(organism, levels = meta_filter$organism %>% unique())) %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  scale_fill_discrete(drop=TRUE,
                      limits = levels(meta_filter$organism)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) + 
  ylab('') + xlab('')

# eye studies
b <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = case_when(Tissue == "Choroid" ~ 'RPE-Choroid',
                            TRUE ~ Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  group_by(Organ, Tissue,organism) %>%
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0)) +
  xlab('')

# eye studies samples
c <-  meta_filter %>%
  mutate(Organ = case_when(Organ == 'Eye' ~ 'Eye',
                           TRUE ~ 'Body'),
         Tissue = gsub('\\.|_', ' ', Tissue),
         Tissue = gsub("RPE$","RPE-Choroid", Tissue),
         Tissue = gsub(" and Aorta","", Tissue)) %>% 
  filter(Organ == 'Eye') %>% 
  select(Organ, Tissue,organism, study_accession) %>%
  unique() %>%
  group_by(Organ, Tissue, organism) %>% 
  summarise(Count = n()) %>%
  ungroup() %>% 
  ggplot(aes(x=Tissue,y=Count, fill = organism, group = organism)) +
  ylab('Number of Studies') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0))  + xlab('')

# retina cell types
d <- meta_filter %>% filter(Organ == 'Eye') %>% 
  mutate(CellType_predict = case_when(is.na(CellType_predict) ~ 'Unlabelled', TRUE ~ CellType_predict)) %>% 
  group_by(CellType_predict, study_accession, organism) %>%
  summarise(Count = n()) %>% filter(Count > 10) %>% ungroup() %>% 
  ggplot(aes(x=CellType_predict,y=Count, fill = organism, group = organism)) +
  ylab('Number of Cells') +
  geom_bar(stat='identity', position = 'dodge') +
  cowplot::theme_cowplot() +
  coord_flip() + 
  scale_y_continuous(expand = c(0, 0)) + xlab('')


a_aa_b <- plot_grid(a,aa,b, ncol = 1, align = 'hv', rel_heights = c(10, 1.5,3.5), labels = 'a')
c_d <- plot_grid(c,d, rel_heights = c(3,10), ncol = 1, align= 'hv', axis = 'l',labels = c('b','c'))
plot_grid(a_aa_b, c_d, ncol = 2, rel_widths = c(1,1.6))
```